---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# GWAS-Engine

<!-- badges: start -->
<!-- badges: end -->

The goal of GWAS-Engine is to provide simple "input/output" style code for GWAS analysis in order to be used inside an API.




# How to use GWAS-Engine


## Dependencies

```{r dep}
# required packages for using GWAS-Engine:
stopifnot(
  compareVersion("1.5.7", as.character(packageVersion("gaston"))) != 1
  )
stopifnot(
  compareVersion("1.7.2", as.character(packageVersion("jsonlite"))) != 1
  )
stopifnot(
  compareVersion("0.3.0", as.character(packageVersion("manhattanly"))) != 1
  )
stopifnot(
  compareVersion("2.5.0", as.character(packageVersion("R6"))) != 1
  )
library(gaston) # for many functions
library(jsonlite) # manage json format
library(manhattanly) # manhattan plot using plotly
library(R6) # R object oriented
```


## Load R scripts

Load all `.R` files in folder `src`.
 
```{r source}
invisible(
  sapply(FUN = source,
         X = list.files("src", pattern = ".R$",full.names = T))
)
```

## Main functions

There is the list of some main function that GWAS-engine can run.

### Run GWAS

```{r runGWAS}
gwas_results <- run_gwas(genoFile = "data/geno/testMarkerData01.vcf.gz",
                         phenoFile = "data/pheno/testPhenoData01.csv",
                         genoUrl = NULL,
                         phenoUrl = NULL,
                         trait = "Flowering.time.at.Arkansas",
                         test = "score",
                         fixed = 0,
                         response = "quantitative",
                         thresh_maf = 0.05,
                         thresh_callrate = 0.95,
                         dir = tempdir())
gwas_results$file
substr(gwas_results$gwasRes, start=1, stop=500)
```


### Draw Manhattan Plot

```{r draw_manhattanPlot}
p <- draw_manhattanPlot(gwasFile = gwas_results$file,
                        gwasUrl = NULL,
                        adj_method = "bonferroni",
                        thresh_p = 0.05,
                        chr = NA)
```
```{r eval=FALSE, include=FALSE}
p
```

![screenshot of the plotly graph](README_files/manPlot.png)

### Adjust p-values

```{r adj_pvals}
gwas_adj <- run_resAdjustment(gwasFile = gwas_results$file,
                              gwasUrl = NULL,
                              adj_method = "bonferroni",
                              dir = tempdir())
substr(gwas_adj$gwasAdjusted, start=1, stop=500)
```


### Draw LD plot

```{r draw_ldPlot}
imgFile <- draw_ldPlot(genoFile = "data/geno/testMarkerData01.vcf.gz",
                       genoUrl = NULL,
                       from = 42,
                       to = 62,
                       dir = tempdir()) 
```

```{r}
library(png)
img <- readPNG(imgFile)
grid::grid.raster(img)
```



## Load data

Example data are stored in the `data` folder.

`readData` read and prepare the data for GWAS analysis
```{r loadDta}
data <- readData(genoFile = "data/geno/testMarkerData01.vcf",
                 phenoFile = "data/pheno/testPhenoData01.csv")
```

For online data, they can be loaded using:

```{r downloadDta, eval=FALSE}
data <- downloadData(genoUrl = "url/to/geno_data.vcf",
                     phenoUrl = "url/to/pheno_data.csv")
```


## GWAS

The `gwas` function return a `data.frame`.

```{r}
gwasRes <- gwas(data = data,
                trait = "Flowering.time.at.Arkansas",
                test = "score",
                fixed = 0,
                response = "quantitative",
                thresh_maf = 0.05,
                thresh_callrate = 0.95)
head(gwasRes)
```


Save GWAS result in a `.json` file.

```{r}
(file <- saveGWAS(gwasRes, metadata = "README"))
cat(paste(readLines(file, n=20), collapse = "\n"))
```

## Manhattan plot

This function generates a "plotly" graph.

```{r manPlot}
p <- manPlot(gwas = gwasRes,
             adj_method = "bonferroni",
             thresh_p = 0.05,
             chr = NA,
             title = "Readme Example")
```
![screenshot of the plotly graph](README_files/manPlot2.png)

```{r eval=FALSE, include=FALSE}
p
```



## LD PLot

Compute $r^2$ Linkage Disequilibrium (LD) between given SNPs and save a plot in a temporary PNG file.

```{r}
gDta <- readGenoData("data/geno/testMarkerData01.vcf.gz")
(imgFile <- LDplot(geno = gDta,
                   from = 1,
                   to = 20,
                   dir = tempdir()))
```


```{r}
img <- readPNG(imgFile)
grid::grid.raster(img)
```


# Functions documentation

All the functions of this engine use `roxygen2` for documentation. You can generate a markdown documentation located in `./doc/README.md` using the function `writeDoc` (defined in `./src/utils.R`):

```{r eval=FALSE}
writeDoc(srcDir = "src",
         docDir = "doc")
```


# Tests

The R package `testthat` is needed to run the unit tests.
To run the unit tests of this engine you can use the command:

```sh
Rscript ./tests/testthat.R
```


# Data references

The genotypic and phenotypic data used as example come from:

Keyan Zhao, Chih-Wei Tung, Georgia C. Eizenga, Mark H. Wright, M. Liakat Ali, Adam H. Price, Gareth J. Norton, M. Rafiqul Islam, Andy Reynolds, Jason Mezey, Anna M. McClung, Carlos D. Bustamante & Susan R. McCouch (2011). [Genome-wide association mapping reveals a rich genetic architecture of complex traits in *Oryza sativa*.](http://www.nature.com/ncomms/journal/v2/n9/full/ncomms1467.html) Nat Comm 2:467 | DOI: 10.1038/ncomms1467, Published Online 13 Sep 2011.
