
-   [r-geno-tools-engine](#r-geno-tools-engine)
-   [Build r-geno-tools-engine](#build-r-geno-tools-engine)
    -   [Test the image](#test-the-image)
-   [How to use r-geno-tools-engine](#how-to-use-r-geno-tools-engine)
    -   [from Docker](#from-docker)
    -   [Without Docker (need R
        installed)](#without-docker-need-r-installed)
        -   [Install dependencies](#install-dependencies)
        -   [usage](#usage)
    -   [As an R library](#as-an-r-library)
-   [Functions documentation](#functions-documentation)
-   [Tests](#tests)
-   [Tools](#tools)
    -   [GWAS](#gwas)
        -   [Introduction](#introduction)
        -   [Functions](#functions)
        -   [Main R functions](#main-r-functions)
    -   [LD plot](#ld-plot)
-   [Data references](#data-references)

<!-- README.md is generated from README.Rmd. Please edit that file -->

# r-geno-tools-engine

<!-- badges: start -->
<!-- badges: end -->

The goal of “r-geno-tools-engine” is to provide a simple “input/output”
style command line toolbox for several genomic related analysis in order
to be used inside an API or other services.

# Build r-geno-tools-engine

The easiest way to use r-geno-tools-engine is to use the
[Docker](https://www.docker.com/) image. To build the docker image
simply do:

``` sh
git clone https://github.com/ut-biomet/r-geno-tools-engine
cd r-geno-tools-engine
docker build -t rgenotoolsengine ./
```

## Test the image

You can test the image by running:

``` sh
docker run --rm --entrypoint="Rscript" rgenotoolsengine ./tests/testthat.R
```

All tests should pass.

# How to use r-geno-tools-engine

## from Docker

Basically the command to use the engine from the docker image is as
follow:

``` sh
docker run --rm rgenotoolsengine <subcommand> --param1 valueOfParam1 --param2 valueOfParam2 ...
```

The sub-command can be one of: (see:
`docker run --rm rgenotoolsengine --help`)

-   `gwas`: to do a gwas analysis from geno and pheno files
-   `gwas-manplot`: to create an interactive or static manhathan plot
    from the results generated by `gwas`
-   `gwas-adjresults`: to calculate the adjusted p-value from the
    results generated by `gwas`
-   `ldplot`: to get a plot showing the Linkage disequilibrium between
    consecutive markers.

The parameters to use depends of the sub-command. You can have an
exhaustive list of them by using:

``` sh
docker run --rm rgenotoolsengine <subcommand> --help
```

However, the tools need to read/write files from the host machine. To
let docker container access those files, you need to mount the the
folders containing the files in the container using the option
`-v --volume`. Then the file path must be specify relative to the place
in the container. (See examples below).

``` sh
docker run --rm \
    -v /path/to/folder1/on/host/:/dockerFolder1 \
    -v /path/to/folder2/on/host/:/dockerFolder2 \
    <subcommand> \
    --fileParam1 "/dockerFolder1/nameOfFile1.txt" \
    --fileParam2 "/dockerFolder2/nameOfFile2.txt" \
    --param3 ...
```

## Without Docker (need R installed)

This r-geno-tools-engine is written in `R`, therefore it can also be run
directly from your computer (without docker) if you have `R` and the all
the dependencies packages installed.

### Install dependencies

To install r-geno-tools-engine’s packages dependencies, first install
the [`renv`](https://rstudio.github.io/renv/articles/renv.html) package.
Then this package will install automatically the dependencies:

In an `R` console:

``` r
# open the repo as an Rstudio project
# or use `setwd('path/to/r-geno-tools-engine)`

# Install `renv`
install.packages(renv)

# Install dependencies
renv::restore()
```

### usage

To use the engine from `R` you can execute the `R-geno-tools-Engine.R`
file with the `Rscript` command.

From a bash terminal

``` sh
Rscript ./r-geno-tools-engine.R <subcommand> --param1 valueOfParam1 --param2 valueOfParam2 ...
```

See:

    Rscript ./r-geno-tools-engine.R --help

and

    Rscript ./r-geno-tools-engine.R <subcommand> --help

## As an R library

It is also possible to use this engine in other project written in `R`
as a library. For that I suggest to include this repository as a git
sub-module of your project.

Then you can load in your code the engine’s function using:

``` r
invisible(
  sapply(FUN = source,
         X = list.files("path/to/r-geno-tools-engine/src", pattern = ".R$",full.names = T))
)
```

For each tool of the engine, some “main functions” are available. Those
functions do essentially the same than the command lines ones. It is
recommended to use those functions in your R project.

Other “utility” functions are also available (eg. for loading data, save
the results…) see the [functions documentation](./doc/README.md) for an
exhaustive list.

# Functions documentation

All the R functions of this engine are documented in
[`./doc/README.md`](doc/README.md).

You can generate this markdown documentation using the function
`writeDoc` (defined in [`./src/utils.R`](src/utils.R)):

``` r
writeDoc(srcDir = "src",
         docDir = "doc")
```

# Tests

The R package `testthat` is needed to run the unit tests. To run the
unit tests of this engine you can use the command:

``` sh
Rscript ./tests/testthat.R
```

# Tools

## GWAS

<details>
<summary>
Click to expand
</summary>

### Introduction

This toolbox provide several genomic analysis. To get a simple
description of each of them, please

GWAS stands for [Genome-Wide Association
Study](https://en.wikipedia.org/wiki/Genome-wide_association_study). It
is a statistical study that aims to detect genetic markers associated
with a particular phenotypic trait.

To test if a marker is associated with a particular phenotypic trait, we
split individuals into different groups according to their genotype for
the studied marker and we test if there is a statistical difference
between the phenotypic values of the groups. The strength of the
statistical difference is measured by what statisticians call the
“[p-value](https://en.wikipedia.org/wiki/P-value)”.

Usually to consider a test “statistically significant” we check if the
p-value of the test is lower than a threshold “α”. In this case the test
is considered positive (*“there is a statistical difference”*) but we
have a probability α to have a false positive. Very often α is equal to
5% or 1%.

GWAS analysis does this test individually and independently for all the
genetic markers in the data-set and records the “statistical
significance” (the p-values) of all the tests.

However, the number of tested genetic markers are usually very large
(several hundreds of thousands or several million) therefore, the number
of false-positive can be high even for a small value of α. For example,
with α=1%, 500000 markers, and if none of them is associated with the
phenotypic trait, we can still expect 5000 “significant test”.

To avoid that, the p-values must be
[adjusted](https://en.wikipedia.org/wiki/Multiple_comparisons_problem)
according to the number of markers.

Finally, the result of a GWAS analysis is presented as a
[Manhattan_plot](https://en.wikipedia.org/wiki/Manhattan_plot) which
shows the `-log(p-value)` on the y-axis and all the markers (order
according to their position on the chromosomes) on the x-axis. A
horizontal line represents the significance threshold and the points of
the markers above the line can be considered as being associated with
the phenotypic trait.

![screenshot of the plotly graph](README_files/manPlot.png)

r-geno-tools-engine provides command line tools to do GWAS analysis.

> Note: You can also check this [7min
> video](https://www.youtube.com/watch?v=sOP8WacfBM8) by Nuno Carvalho
> that explain GWAS very well.

### Functions

#### `gwas`

To do a gwas analysis and write the results in a json file.

``` sh
docker run --rm -v "$PWD"/data/geno/:/geno \
    -v "$PWD"/data/pheno/:/pheno \
    -v "$PWD"/readmeTemp:/out rgenotoolsengine \
    gwas \
    --genoFile "/geno/testMarkerData01.vcf.gz" \
    --phenoFile "/pheno/testPhenoData01.csv" \
    --trait "Flowering.time.at.Arkansas" \
    --test "score" \
    --fixed 0 \
    --response "quantitative" \
    --thresh_maf 0.05 \
    --thresh_callrate 0.95 \
    --outFile "/out/gwasRes.json"
```

#### `gwas-manplot`

To create a Manhattan plot.

**Interactive plot:**

``` sh
docker run --rm -v "$PWD"/readmeTemp:/files rgenotoolsengine \
    gwas-manplot \
    --gwasFile "/files/gwasRes.json" \
    --adj_method "bonferroni" \
    --thresh_p 0.05 \
    --interactive TRUE \
    --filter_nPoints 3000 \
    --outFile "/files/manPlot.html"
```

**Static plot:**

``` sh
docker run --rm -v "$PWD"/readmeTemp:/files rgenotoolsengine \
    gwas-manplot \
    --gwasFile "/files/gwasRes.json" \
    --adj_method "bonferroni" \
    --thresh_p 0.05 \
    --interactive FALSE \
    --outFile "/files/manPlot.png"
```

#### `adjresults`

To create a `json` file with the gwas results with adjusted p-values. It
is also possible to filter the result to keep the values with the lowest
p-values.

``` sh
docker run --rm -v "$PWD"/readmeTemp:/files rgenotoolsengine \
    gwas-adjResults \
    --gwasFile "/files/gwasRes.json" \
    --adj_method "bonferroni" \
    --filter_nPoints 3000 \
    --outFile "/files/adjRes.json"
```

### Main R functions

For a usage as R library one can use those functions.

``` r
gwas_results <- run_gwas(genoFile = "data/geno/testMarkerData01.vcf.gz",
                         phenoFile = "data/pheno/testPhenoData01.csv",
                         genoUrl = NULL,
                         phenoUrl = NULL,
                         trait = "Flowering.time.at.Arkansas",
                         test = "score",
                         fixed = 0,
                         response = "quantitative",
                         thresh_maf = 0.05,
                         thresh_callrate = 0.95,
                         outFile = tempfile(fileext = ".json"))
#> 2021-12-07 15:51:00 - r-run_gwas(): Get data ...
#> 2021-12-07 15:51:00 - r-readData(): get geno data ...
#> 2021-12-07 15:51:00 - r-readGenoData(): Check file extention ... 
#> 2021-12-07 15:51:00 - r-readGenoData(): Read geno file ... 
#> ped stats and snps stats have been set. 
#> 'p' has been set. 
#> 'mu' and 'sigma' have been set.
#> 2021-12-07 15:51:01 - r-readGenoData(): Read geno file DONE 
#> 2021-12-07 15:51:01 - r-readGenoData(): DONE, return output.
#> 2021-12-07 15:51:01 - r-readData(): get geno data DONE
#> 2021-12-07 15:51:01 - r-readData(): get pheno data ...
#> 2021-12-07 15:51:01 - r-readPhenoData(): Read phenotypic file ... 
#> 2021-12-07 15:51:01 - r-readPhenoData(): Read phenotypic file DONE 
#> 2021-12-07 15:51:01 - r-readPhenoData(): DONE, return output.
#> 2021-12-07 15:51:01 - r-readData(): get pheno data DONE
#> 2021-12-07 15:51:01 - r-readData(): prepare data ...
#> 2021-12-07 15:51:01 - r-prepareData(): Remove from geno data individuals that are not in phenotypic data-set ...
#> 2021-12-07 15:51:01 - r-prepareData(): Remove from geno data individuals that are not in phenotypic data-set DONE
#> 2021-12-07 15:51:01 - r-prepareData(): reorder matrix ...
#> 2021-12-07 15:51:01 - r-prepareData(): reorder matrix DONE
#> 2021-12-07 15:51:01 - r-prepareData(): remove monomorphic markers ...
#> 2021-12-07 15:51:01 - r-prepareData(): remove monomorphic markers DONE
#> 2021-12-07 15:51:01 - r-prepareData(): DONE, return output.
#> 2021-12-07 15:51:01 - r-readData(): prepare data DONE
#> 2021-12-07 15:51:01 - r-readData(): DONE, return output.
#> 2021-12-07 15:51:01 - r-run_gwas(): Get data DONE
#> 2021-12-07 15:51:01 - r-run_gwas(): GWAS analysis ...
#> 2021-12-07 15:51:01 - r-gwas(): Check inputs ...
#> 2021-12-07 15:51:01 - r-gwas(): Check inputs DONE
#> 2021-12-07 15:51:01 - r-gwas(): aggregate data in bed matrix ...
#> 2021-12-07 15:51:01 - r-gwas(): aggregate data in bed matrix DONE
#> 2021-12-07 15:51:01 - r-gwas(): remove individuals with missing phenotypic values ...
#> 2021-12-07 15:51:02 - r-gwas(): remove samples with missing phenotypic values DONE
#> 2021-12-07 15:51:02 - r-gwas(): filter SNPs ...
#> 2021-12-07 15:51:02 - r-gwas(): filter SNPs DONE
#> 2021-12-07 15:51:02 - r-gwas(): calculate genetic relatinoal matrix ...
#> 2021-12-07 15:51:02 - r-gwas(): calculate genetic relatinoal matrix DONE
#> 2021-12-07 15:51:02 - r-gwas(): fit model ...
#> [Iteration 1] theta = 78.0648  78.282
#> [Iteration 1] log L = -1002.17
#> [Iteration 1] AI-REML update
#> [Iteration 1] ||gradient|| = 0.344543
#> [Iteration 2] theta = 22.5611 163.026
#> [Iteration 2] log L = -990.943
#> [Iteration 2] AI-REML update
#> [Iteration 2] ||gradient|| = 0.798817
#> [Iteration 3] theta =   27.72 190.624
#> [Iteration 3] log L = -986.649
#> [Iteration 3] AI-REML update
#> [Iteration 3] ||gradient|| = 0.113945
#> [Iteration 4] theta = 29.2503 195.436
#> [Iteration 4] log L = -986.514
#> [Iteration 4] AI-REML update
#> [Iteration 4] ||gradient|| = 0.00580941
#> [Iteration 5] theta = 29.4021 195.213
#> [Iteration 5] log L = -986.513
#> [Iteration 5] AI-REML update
#> [Iteration 5] ||gradient|| = 0.000323293
#> [Iteration 6] theta = 29.4178 195.142
#> [Iteration 6] log L = -986.513
#> [Iteration 6] AI-REML update
#> [Iteration 6] ||gradient|| = 4.3263e-05
#> [Iteration 7] theta = 29.4199 195.133
#> [Iteration 7] log L = -986.513
#> [Iteration 7] AI-REML update
#> [Iteration 7] ||gradient|| = 5.95127e-06
#> 2021-12-07 15:51:02 - r-gwas(): fit model DONE
#> 2021-12-07 15:51:02 - r-gwas(): DONE, return output.
#> 2021-12-07 15:51:02 - r-run_gwas(): GWAS analysis DONE
#> 2021-12-07 15:51:02 - r-run_gwas(): Save metadata ...
#> 2021-12-07 15:51:02 - r-run_gwas(): Save metadata DONE
#> 2021-12-07 15:51:02 - r-run_gwas(): Save results ...
#> 2021-12-07 15:51:02 - r-saveGWAS(): Check file ...
#> 2021-12-07 15:51:02 - r-saveGWAS(): Check file DONE
#> 2021-12-07 15:51:03 - r-run_gwas(): Save results DONE
gwas_results$file
#> [1] "/tmp/RtmppUbRTz/file8c0d793ac574.json"
substr(gwas_results$gwasRes, start=1, stop=500)
#> [
#>   {
#>     "chr": "1",
#>     "pos": 9563,
#>     "id": "SNP-1.8562.",
#>     "A1": "A",
#>     "A2": "T",
#>     "freqA2": 0.1359,
#>     "score": 4.395,
#>     "p": 0.036
#>   },
#>   {
#>     "chr": "1",
#>     "pos": 25922,
#>     "id": "SNP-1.24921.",
#>     "A1": "C",
#>     "A2": "T",
#>     "freqA2": 0.1254,
#>     "score": 1.6745,
#>     "p": 0.1957
#>   },
#>   {
#>     "chr": "1",
#>     "pos": 26254,
#>     "id": "SNP-1.25253.",
#>     "A1": "A",
#>     "A2": "T",
#>     "freqA2": 0.2935,
#>     "score": 3.6592,
#>     "p": 0.0558
#>   },
#>   {
#>     "chr": "1",
#>     "pos
```

``` r
p <- draw_manhattanPlot(gwasFile = gwas_results$file,
                        gwasUrl = NULL,
                        adj_method = "bonferroni",
                        thresh_p = 0.05,
                        chr = NA,
                        interactive = TRUE,
                        # filter_pAdj = 1,
                        # filter_nPoints = Inf,
                        filter_quant = 0.1,
                        outFile = tempfile(fileext = ".html"))
#> 2021-12-07 15:51:03 - r-draw_manhattanPlot(): Check outFile ...
#> 2021-12-07 15:51:03 - r-draw_manhattanPlot(): Check outFile DONE
#> 2021-12-07 15:51:03 - r-draw_manhattanPlot(): Get data ...
#> 2021-12-07 15:51:03 - r-readGWAS(): Read result file ... 
#> 2021-12-07 15:51:03 - r-readGWAS(): Read result file DONE 
#> 2021-12-07 15:51:03 - r-readGWAS(): Convert Json to data.frame ... 
#> 2021-12-07 15:51:03 - r-readGWAS(): Convert Json to data.frame DONE 
#> 2021-12-07 15:51:03 - r-readGWAS(): DONE, return output.
#> 2021-12-07 15:51:03 - r-draw_manhattanPlot(): Get data DONE
#> 2021-12-07 15:51:03 - r-draw_manhattanPlot(): Draw Manhattan Plot ...
#> 2021-12-07 15:51:03 - r-manPlot(): Check parameters...
#> 2021-12-07 15:51:03 - r-manPlot(): Check parameters DONE
#> 2021-12-07 15:51:03 - r-manPlot(): Check chromosome name ...
#> 2021-12-07 15:51:03 - r-manPlot(): Check chromosome name DONE
#> 2021-12-07 15:51:03 - r-manPlot(): Remove NAs ...
#> 2021-12-07 15:51:03 - r-manPlot(): Remove NAs DONE
#> 2021-12-07 15:51:03 - r-manPlot(): Adjust p-values ...
#> 2021-12-07 15:51:03 - r-adjustPval(): Check adj_method ...
#> 2021-12-07 15:51:03 - r-adjustPval(): Check adj_method DONE
#> 2021-12-07 15:51:03 - r-adjustPval(): Check p values ...
#> 2021-12-07 15:51:03 - r-adjustPval(): Check p values DONE
#> 2021-12-07 15:51:03 - r-adjustPval(): Adjust p-values ...
#> 2021-12-07 15:51:03 - r-adjustPval(): Adjust p-values DONE
#> 2021-12-07 15:51:03 - r-adjustPval(): Adjust threshold ...
#> 2021-12-07 15:51:04 - r-adjustPval(): Adjust threshold DONE
#> 2021-12-07 15:51:04 - r-adjustPval(): DONE, return output
#> 2021-12-07 15:51:04 - r-manPlot(): Adjust p-values DONE
#> 2021-12-07 15:51:04 - r-manPlot(): Check duplicated SNP ID ...
#> 2021-12-07 15:51:04 - r-manPlot(): Check duplicated SNP ID DONE
#> 2021-12-07 15:51:04 - r-manPlot(): Extract significant SNP ...
#> 2021-12-07 15:51:04 - r-manPlot(): Extract significant SNP DONE
#> 2021-12-07 15:51:04 - r-filterGWAS(): Check parameters...
#> 2021-12-07 15:51:04 - r-filterGWAS(): Check parameters DONE
#> 2021-12-07 15:51:04 - r-filterGWAS(): Filter points ...
#> 2021-12-07 15:51:04 - r-filterGWAS(): skip filter_pAdj
#> 2021-12-07 15:51:04 - r-filterGWAS(): skip filter_nPoints
#> 2021-12-07 15:51:04 - r-manPlot(): Draw plot ...
#> 2021-12-07 15:51:04 - r-manPlot(): Draw plot DONE
#> 2021-12-07 15:51:04 - r-manPlot(): DONE, return output
#> 2021-12-07 15:51:04 - r-draw_manhattanPlot(): Draw Manhattan Plot DONE
#> 2021-12-07 15:51:04 - r-draw_manhattanPlot(): Save results ...
#> 2021-12-07 15:51:06 - r-draw_manhattanPlot(): Save results DONE
```

``` r
gwas_adj <- run_resAdjustment(gwasFile = gwas_results$file,
                              gwasUrl = NULL,
                              adj_method = "bonferroni",
                              outFile = tempfile(fileext = ".json"))
#> 2021-12-07 15:51:06 - r-run_resAdjustment(): Get data ...
#> 2021-12-07 15:51:06 - r-readGWAS(): Read result file ... 
#> 2021-12-07 15:51:06 - r-readGWAS(): Read result file DONE 
#> 2021-12-07 15:51:06 - r-readGWAS(): Convert Json to data.frame ... 
#> 2021-12-07 15:51:07 - r-readGWAS(): Convert Json to data.frame DONE 
#> 2021-12-07 15:51:07 - r-readGWAS(): DONE, return output.
#> 2021-12-07 15:51:07 - r-run_resAdjustment(): Get data DONE
#> 2021-12-07 15:51:07 - r-run_resAdjustment(): Adjust p-values ...
#> 2021-12-07 15:51:07 - r-adjustPval(): Check adj_method ...
#> 2021-12-07 15:51:07 - r-adjustPval(): Check adj_method DONE
#> 2021-12-07 15:51:07 - r-adjustPval(): Check p values ...
#> 2021-12-07 15:51:07 - r-adjustPval(): Check p values DONE
#> 2021-12-07 15:51:07 - r-adjustPval(): Adjust p-values ...
#> 2021-12-07 15:51:07 - r-adjustPval(): Adjust p-values DONE
#> 2021-12-07 15:51:07 - r-adjustPval(): DONE, return output
#> 2021-12-07 15:51:07 - r-run_resAdjustment(): Adjust p-values DONE
#> 2021-12-07 15:51:07 - r-filterGWAS(): Check parameters...
#> 2021-12-07 15:51:07 - r-filterGWAS(): Check parameters DONE
#> 2021-12-07 15:51:07 - r-filterGWAS(): Filter points ...
#> 2021-12-07 15:51:07 - r-filterGWAS(): skip filter_pAdj
#> 2021-12-07 15:51:07 - r-filterGWAS(): skip filter_quant
#> 2021-12-07 15:51:07 - r-filterGWAS(): skip filter_nPoints
#> 2021-12-07 15:51:07 - r-run_resAdjustment(): Save results ...
#> 2021-12-07 15:51:07 - r-saveGWAS(): Check file ...
#> 2021-12-07 15:51:07 - r-saveGWAS(): Check file DONE
#> 2021-12-07 15:51:07 - r-run_resAdjustment(): Save results DONE
substr(gwas_adj$gwasAdjusted, start=1, stop=500)
#> [
#>   {
#>     "chr": "1",
#>     "pos": 9563,
#>     "id": "SNP-1.8562.",
#>     "A1": "A",
#>     "A2": "T",
#>     "freqA2": 0.1359,
#>     "score": 4.395,
#>     "p": 0.036,
#>     "p_adj": 1
#>   },
#>   {
#>     "chr": "1",
#>     "pos": 25922,
#>     "id": "SNP-1.24921.",
#>     "A1": "C",
#>     "A2": "T",
#>     "freqA2": 0.1254,
#>     "score": 1.6745,
#>     "p": 0.1957,
#>     "p_adj": 1
#>   },
#>   {
#>     "chr": "1",
#>     "pos": 26254,
#>     "id": "SNP-1.25253.",
#>     "A1": "A",
#>     "A2": "T",
#>     "freqA2": 0.2935,
#>     "score": 3.6592,
#>     "p": 0.0558,
#> 
```

</details>

## LD plot

<details>
<summary>
Click to expand
</summary>

This help to calculate and visualize the linkage disequilibrium between
some consecutive snps.

``` sh
docker run --rm -v "$PWD"/data/geno/:/geno \
    -v "$PWD"/readmeTemp:/out rgenotoolsengine \
    ldplot \
    --genoFile "/geno/testMarkerData01.vcf.gz" \
    --from 42 \
    --to 62 \
    --outFile "/out/ldplot.png"
```

**Main function**

``` r
imgFile <- draw_ldPlot(genoFile = "data/geno/testMarkerData01.vcf.gz",
                       genoUrl = NULL,
                       from = 42,
                       to = 62,
                       outFile = tempfile(fileext = ".png")) 
#> 2021-12-07 15:51:07 - r-draw_ldPlot(): Get data ...
#> 2021-12-07 15:51:07 - r-readGenoData(): Check file extention ... 
#> 2021-12-07 15:51:07 - r-readGenoData(): Read geno file ... 
#> ped stats and snps stats have been set. 
#> 'p' has been set. 
#> 'mu' and 'sigma' have been set.
#> 2021-12-07 15:51:08 - r-readGenoData(): Read geno file DONE 
#> 2021-12-07 15:51:08 - r-readGenoData(): DONE, return output.
#> 2021-12-07 15:51:08 - r-draw_ldPlot(): Get data DONE
#> 2021-12-07 15:51:08 - r-draw_ldPlot(): Draw LD Plot ...
#> 2021-12-07 15:51:08 - r-LDplot(): Check "from" and "to" format ...
#> 2021-12-07 15:51:08 - r-LDplot(): Check "from" and "to" format DONE
#> 2021-12-07 15:51:08 - r-LDplot(): Check "from" < "to"...
#> 2021-12-07 15:51:08 - r-LDplot(): Check "from" < "to" DONE
#> 2021-12-07 15:51:08 - r-LDplot(): Check number of SNP < 50...
#> 2021-12-07 15:51:08 - r-LDplot(): Check number of SNP < 50 DONE
#> 2021-12-07 15:51:08 - r-LDplot(): Check file ...
#> 2021-12-07 15:51:08 - r-LDplot(): Check file DONE
#> 2021-12-07 15:51:08 - r-LDplot(): Compute LD ...
#> 2021-12-07 15:51:08 - r-LDplot(): Compute LD DONE
#> 2021-12-07 15:51:08 - r-LDplot(): Create LD plot ...
#> 2021-12-07 15:51:08 - r-LDplot(): Create create file: /tmp/RtmppUbRTz/file8c0d5088477a.png
#> 2021-12-07 15:51:08 - r-LDplot(): Create LD plot DONE
#> 2021-12-07 15:51:08 - r-LDplot(): DONE, return output
#> 2021-12-07 15:51:08 - r-draw_ldPlot(): Draw LD Plot DONE
```

</details>

# Data references

Some example data are available in the folder [data](data/). This folder
also includes examples of output files.

The genotypic and phenotypic data used as example come from:

> Keyan Zhao, Chih-Wei Tung, Georgia C. Eizenga, Mark H. Wright, M.
> Liakat Ali, Adam H. Price, Gareth J. Norton, M. Rafiqul Islam, Andy
> Reynolds, Jason Mezey, Anna M. McClung, Carlos D. Bustamante & Susan
> R. McCouch (2011). [Genome-wide association mapping reveals a rich
> genetic architecture of complex traits in *Oryza
> sativa*.](http://www.nature.com/ncomms/journal/v2/n9/full/ncomms1467.html)
> Nat Comm 2:467 \| DOI: 10.1038/ncomms1467, Published Online 13 Sep
> 2011.
