
-   [r-geno-tools-engine](#r-geno-tools-engine)
-   [Build r-geno-tools-engine](#build-r-geno-tools-engine)
    -   [Test the image](#test-the-image)
-   [How to use r-geno-tools-engine](#how-to-use-r-geno-tools-engine)
    -   [from Docker](#from-docker)
    -   [Without Docker (need R
        installed)](#without-docker-need-r-installed)
        -   [Install dependencies](#install-dependencies)
        -   [usage](#usage)
    -   [As an R library](#as-an-r-library)
-   [Functions documentation](#functions-documentation)
-   [Tests](#tests)
-   [Tools](#tools)
    -   [GWAS](#gwas)
        -   [Introduction](#introduction)
        -   [Functions](#functions)
        -   [Main R functions](#main-r-functions)
    -   [LD plot](#ld-plot)
    -   [Relationship matrix](#relationship-matrix)
        -   [Pedigree relationship
            matrix](#pedigree-relationship-matrix)
        -   [Relationship matrix
            visualisation](#relationship-matrix-visualisation)
    -   [Pedigree network](#pedigree-network)
-   [Data references](#data-references)
-   [Utils](#utils)

<!-- README.md is generated from README.Rmd. Please edit that file -->

# r-geno-tools-engine

<!-- badges: start -->
<!-- badges: end -->

The goal of “r-geno-tools-engine” is to provide a simple “input/output”
style command line toolbox for several genomic related analysis in order
to be used inside an API or other services.

# Build r-geno-tools-engine

The easiest way to use r-geno-tools-engine is to use the
[Docker](https://www.docker.com/) image. To build the docker image
simply do:

``` sh
git clone https://github.com/ut-biomet/r-geno-tools-engine
cd r-geno-tools-engine
docker build -t rgenotoolsengine ./
```

## Test the image

You can test the image by running:

``` sh
docker run --rm --entrypoint="Rscript" rgenotoolsengine ./tests/testthat.R
```

All tests should pass.

# How to use r-geno-tools-engine

## from Docker

Basically the command to use the engine from the docker image is as
follow:

``` sh
docker run --rm rgenotoolsengine <subcommand> --param1 valueOfParam1 --param2 valueOfParam2 ...
```

The sub-command can be one of: (see:
`docker run --rm rgenotoolsengine --help`)

-   `gwas`: to do a gwas analysis from geno and pheno files
-   `gwas-manplot`: to create an interactive or static manhathan plot
    from the results generated by `gwas`
-   `gwas-adjresults`: to calculate the adjusted p-value from the
    results generated by `gwas`
-   `ldplot`: to get a plot showing the Linkage disequilibrium between
    consecutive markers.
-   `relmat-ped`: to calculate the relationship matrix from a pedigree
    file.
-   `relmat-heatmap`: to get the heatmap of a relationship matrix.
-   `pedNetwork`: to get an interactive pedigree network from a pedigree
    file.

The parameters to use depends of the sub-command. You can have an
exhaustive list of them by using:

``` sh
docker run --rm rgenotoolsengine <subcommand> --help
```

However, the tools need to read/write files from the host machine. To
let docker container access those files, you need to mount the the
folders containing the files in the container using the option
`-v --volume`. Then the file path must be specify relative to the place
in the container. (See examples below).

``` sh
docker run --rm \
    -v /path/to/folder1/on/host/:/dockerFolder1 \
    -v /path/to/folder2/on/host/:/dockerFolder2 \
    <subcommand> \
    --fileParam1 "/dockerFolder1/nameOfFile1.txt" \
    --fileParam2 "/dockerFolder2/nameOfFile2.txt" \
    --param3 ...
```

## Without Docker (need R installed)

This r-geno-tools-engine is written in `R`, therefore it can also be run
directly from your computer (without docker) if you have `R` and the all
the dependencies packages installed.

### Install dependencies

To install r-geno-tools-engine’s packages dependencies, first install
the [`renv`](https://rstudio.github.io/renv/articles/renv.html) package.
Then this package will install automatically the dependencies:

In an `R` console:

``` r
# open the repo as an Rstudio project
# or use `setwd('path/to/r-geno-tools-engine)`

# Install `renv`
install.packages(renv)

# Install dependencies
renv::restore()
```

### usage

To use the engine from `R` you can execute the `R-geno-tools-Engine.R`
file with the `Rscript` command.

From a bash terminal

``` sh
Rscript ./r-geno-tools-engine.R <subcommand> --param1 valueOfParam1 --param2 valueOfParam2 ...
```

See:

    Rscript ./r-geno-tools-engine.R --help

and

    Rscript ./r-geno-tools-engine.R <subcommand> --help

## As an R library

It is also possible to use this engine in other project written in `R`
as a library. For that I suggest to include this repository as a git
sub-module of your project.

Then you can load in your code the engine’s function using:

``` r
invisible(
  sapply(FUN = source,
         X = list.files("path/to/r-geno-tools-engine/src", pattern = ".R$",full.names = T))
)
```

For each tool of the engine, some “main functions” are available. Those
functions do essentially the same than the command lines ones. It is
recommended to use those functions in your R project.

Other “utility” functions are also available (eg. for loading data, save
the results…) see the [functions documentation](./doc/README.md) for an
exhaustive list.

# Functions documentation

All the R functions of this engine are documented in
[`./doc/README.md`](doc/README.md).

You can generate this markdown documentation using the function
`writeDoc` (defined in [`./src/utils.R`](src/utils.R)):

``` r
writeDoc(srcDir = "src",
         docDir = "doc")
```

# Tests

The R package `testthat` is needed to run the unit tests. To run the
unit tests of this engine you can use the command:

``` sh
Rscript ./tests/testthat.R
```

# Tools

## GWAS

<details>
<summary>
Click to expand
</summary>

### Introduction

This toolbox provide several genomic analysis. To get a simple
description of each of them, please

GWAS stands for [Genome-Wide Association
Study](https://en.wikipedia.org/wiki/Genome-wide_association_study). It
is a statistical study that aims to detect genetic markers associated
with a particular phenotypic trait.

To test if a marker is associated with a particular phenotypic trait, we
split individuals into different groups according to their genotype for
the studied marker and we test if there is a statistical difference
between the phenotypic values of the groups. The strength of the
statistical difference is measured by what statisticians call the
“[p-value](https://en.wikipedia.org/wiki/P-value)”.

Usually to consider a test “statistically significant” we check if the
p-value of the test is lower than a threshold “α”. In this case the test
is considered positive (*“there is a statistical difference”*) but we
have a probability α to have a false positive. Very often α is equal to
5% or 1%.

GWAS analysis does this test individually and independently for all the
genetic markers in the data-set and records the “statistical
significance” (the p-values) of all the tests.

However, the number of tested genetic markers are usually very large
(several hundreds of thousands or several million) therefore, the number
of false-positive can be high even for a small value of α. For example,
with α=1%, 500000 markers, and if none of them is associated with the
phenotypic trait, we can still expect 5000 “significant test”.

To avoid that, the p-values must be
[adjusted](https://en.wikipedia.org/wiki/Multiple_comparisons_problem)
according to the number of markers.

Finally, the result of a GWAS analysis is presented as a
[Manhattan_plot](https://en.wikipedia.org/wiki/Manhattan_plot) which
shows the `-log(p-value)` on the y-axis and all the markers (order
according to their position on the chromosomes) on the x-axis. A
horizontal line represents the significance threshold and the points of
the markers above the line can be considered as being associated with
the phenotypic trait.

![screenshot of the plotly graph](README_files/manPlot.png)

r-geno-tools-engine provides command line tools to do GWAS analysis.

> Note: You can also check this [7min
> video](https://www.youtube.com/watch?v=sOP8WacfBM8) by Nuno Carvalho
> that explain GWAS very well.

### Functions

#### `gwas`

To do a gwas analysis and write the results in a json file.

``` sh
docker run --rm -v "$PWD"/data/geno/:/geno \
    -v "$PWD"/data/pheno/:/pheno \
    -v "$PWD"/readmeTemp:/out rgenotoolsengine \
    gwas \
    --genoFile "/geno/testMarkerData01.vcf.gz" \
    --phenoFile "/pheno/testPhenoData01.csv" \
    --trait "Flowering.time.at.Arkansas" \
    --test "score" \
    --fixed 0 \
    --response "quantitative" \
    --thresh_maf 0.05 \
    --thresh_callrate 0.95 \
    --outFile "/out/gwasRes.json"
```

#### `gwas-manplot`

To create a Manhattan plot.

**Interactive plot:**

``` sh
docker run --rm -v "$PWD"/readmeTemp:/files rgenotoolsengine \
    gwas-manplot \
    --gwasFile "/files/gwasRes.json" \
    --adj_method "bonferroni" \
    --thresh_p 0.05 \
    --interactive TRUE \
    --filter_nPoints 3000 \
    --outFile "/files/manPlot.html"
```

**Static plot:**

``` sh
docker run --rm -v "$PWD"/readmeTemp:/files rgenotoolsengine \
    gwas-manplot \
    --gwasFile "/files/gwasRes.json" \
    --adj_method "bonferroni" \
    --thresh_p 0.05 \
    --interactive FALSE \
    --outFile "/files/manPlot.png"
```

#### `adjresults`

To create a `json` file with the gwas results with adjusted p-values. It
is also possible to filter the result to keep the values with the lowest
p-values.

``` sh
docker run --rm -v "$PWD"/readmeTemp:/files rgenotoolsengine \
    gwas-adjResults \
    --gwasFile "/files/gwasRes.json" \
    --adj_method "bonferroni" \
    --filter_nPoints 3000 \
    --outFile "/files/adjRes.json"
```

### Main R functions

For a usage as R library one can use those functions.

``` r
gwas_results <- run_gwas(genoFile = "data/geno/testMarkerData01.vcf.gz",
                         phenoFile = "data/pheno/testPhenoData01.csv",
                         genoUrl = NULL,
                         phenoUrl = NULL,
                         trait = "Flowering.time.at.Arkansas",
                         test = "score",
                         fixed = 0,
                         response = "quantitative",
                         thresh_maf = 0.05,
                         thresh_callrate = 0.95,
                         outFile = tempfile(fileext = ".json"))
#> 2022-04-11 15:02:24 - r-run_gwas(): Get data ...
#> 2022-04-11 15:02:24 - r-readData(): get geno data ...
#> 2022-04-11 15:02:24 - r-readGenoData(): Check file extention ... 
#> 2022-04-11 15:02:24 - r-readGenoData(): Read geno file ... 
#> ped stats and snps stats have been set. 
#> 'p' has been set. 
#> 'mu' and 'sigma' have been set.
#> 2022-04-11 15:02:25 - r-readGenoData(): Read geno file DONE 
#> 2022-04-11 15:02:25 - r-readGenoData(): DONE, return output.
#> 2022-04-11 15:02:25 - r-readData(): get geno data DONE
#> 2022-04-11 15:02:25 - r-readData(): get pheno data ...
#> 2022-04-11 15:02:25 - r-readPhenoData(): Read phenotypic file ... 
#> 2022-04-11 15:02:25 - r-readPhenoData(): Read phenotypic file DONE 
#> 2022-04-11 15:02:25 - r-readPhenoData(): Check individuals unicity ...
#> 2022-04-11 15:02:25 - r-readPhenoData(): Check individuals unicity DONE
#> 2022-04-11 15:02:25 - r-readPhenoData(): Set pheno data's row names ...
#> 2022-04-11 15:02:25 - r-readPhenoData(): Set pheno data's row names DONE
#> 2022-04-11 15:02:25 - r-readPhenoData(): DONE, return output.
#> 2022-04-11 15:02:25 - r-readData(): get pheno data DONE
#> 2022-04-11 15:02:25 - r-readData(): prepare data ...
#> 2022-04-11 15:02:25 - r-prepareData(): Remove from geno data individuals that are not in phenotypic data-set ...
#> 2022-04-11 15:02:25 - r-prepareData(): Remove from geno data individuals that are not in phenotypic data-set DONE
#> 2022-04-11 15:02:25 - r-prepareData(): reorder matrix ...
#> 2022-04-11 15:02:25 - r-prepareData(): reorder matrix DONE
#> 2022-04-11 15:02:25 - r-prepareData(): remove monomorphic markers ...
#> 2022-04-11 15:02:25 - r-prepareData(): remove monomorphic markers DONE
#> 2022-04-11 15:02:25 - r-prepareData(): DONE, return output.
#> 2022-04-11 15:02:25 - r-readData(): prepare data DONE
#> 2022-04-11 15:02:25 - r-readData(): DONE, return output.
#> 2022-04-11 15:02:25 - r-run_gwas(): Get data DONE
#> 2022-04-11 15:02:25 - r-run_gwas(): GWAS analysis ...
#> 2022-04-11 15:02:25 - r-gwas(): Check inputs ...
#> 2022-04-11 15:02:25 - r-gwas(): Check inputs DONE
#> 2022-04-11 15:02:25 - r-gwas(): aggregate data in bed matrix ...
#> 2022-04-11 15:02:25 - r-gwas(): aggregate data in bed matrix DONE
#> 2022-04-11 15:02:25 - r-gwas(): remove individuals with missing phenotypic values ...
#> 2022-04-11 15:02:26 - r-gwas(): remove samples with missing phenotypic values DONE
#> 2022-04-11 15:02:26 - r-gwas(): filter SNPs ...
#> 2022-04-11 15:02:26 - r-gwas(): filter SNPs DONE
#> 2022-04-11 15:02:26 - r-gwas(): calculate genetic relatinoal matrix ...
#> 2022-04-11 15:02:26 - r-gwas(): calculate genetic relatinoal matrix DONE
#> 2022-04-11 15:02:26 - r-gwas(): fit model ...
#> [Iteration 1] theta = 78.0648 78.2819
#> [Iteration 1] log L = -1002.17
#> [Iteration 1] AI-REML update
#> [Iteration 1] ||gradient|| = 0.344543
#> [Iteration 2] theta =  22.561 163.026
#> [Iteration 2] log L = -990.943
#> [Iteration 2] AI-REML update
#> [Iteration 2] ||gradient|| = 0.79882
#> [Iteration 3] theta = 27.7199 190.624
#> [Iteration 3] log L = -986.649
#> [Iteration 3] AI-REML update
#> [Iteration 3] ||gradient|| = 0.113946
#> [Iteration 4] theta = 29.2502 195.436
#> [Iteration 4] log L = -986.513
#> [Iteration 4] AI-REML update
#> [Iteration 4] ||gradient|| = 0.0058095
#> [Iteration 5] theta =  29.402 195.213
#> [Iteration 5] log L = -986.513
#> [Iteration 5] AI-REML update
#> [Iteration 5] ||gradient|| = 0.000323301
#> [Iteration 6] theta = 29.4177 195.143
#> [Iteration 6] log L = -986.513
#> [Iteration 6] AI-REML update
#> [Iteration 6] ||gradient|| = 4.32643e-05
#> [Iteration 7] theta = 29.4199 195.133
#> [Iteration 7] log L = -986.513
#> [Iteration 7] AI-REML update
#> [Iteration 7] ||gradient|| = 5.95149e-06
#> 2022-04-11 15:02:26 - r-gwas(): fit model DONE
#> 2022-04-11 15:02:26 - r-gwas(): DONE, return output.
#> 2022-04-11 15:02:26 - r-run_gwas(): GWAS analysis DONE
#> 2022-04-11 15:02:26 - r-run_gwas(): Save metadata ...
#> 2022-04-11 15:02:26 - r-run_gwas(): Save metadata DONE
#> 2022-04-11 15:02:26 - r-run_gwas(): Save results ...
#> 2022-04-11 15:02:26 - r-saveGWAS(): Check file ...
#> 2022-04-11 15:02:26 - r-saveGWAS(): Check file DONE
#> 2022-04-11 15:02:26 - r-run_gwas(): Save results DONE
gwas_results$file
#> [1] "/tmp/Rtmpi6YPG5/file1e198a654c66b1.json"
substr(gwas_results$gwasRes, start=1, stop=500)
#> [
#>   {
#>     "chr": "1",
#>     "pos": 9563,
#>     "id": "SNP-1.8562.",
#>     "A1": "A",
#>     "A2": "T",
#>     "freqA2": 0.1359,
#>     "score": 4.395,
#>     "p": 0.036
#>   },
#>   {
#>     "chr": "1",
#>     "pos": 25922,
#>     "id": "SNP-1.24921.",
#>     "A1": "C",
#>     "A2": "T",
#>     "freqA2": 0.1254,
#>     "score": 1.6744,
#>     "p": 0.1957
#>   },
#>   {
#>     "chr": "1",
#>     "pos": 26254,
#>     "id": "SNP-1.25253.",
#>     "A1": "A",
#>     "A2": "T",
#>     "freqA2": 0.2935,
#>     "score": 3.6592,
#>     "p": 0.0558
#>   },
#>   {
#>     "chr": "1",
#>     "pos
```

``` r
p <- draw_manhattanPlot(gwasFile = gwas_results$file,
                        gwasUrl = NULL,
                        adj_method = "bonferroni",
                        thresh_p = 0.05,
                        chr = NA,
                        interactive = TRUE,
                        # filter_pAdj = 1,
                        # filter_nPoints = Inf,
                        filter_quant = 0.1,
                        outFile = tempfile(fileext = ".html"))
#> 2022-04-11 15:02:27 - r-draw_manhattanPlot(): Check outFile ...
#> 2022-04-11 15:02:27 - r-draw_manhattanPlot(): Check outFile DONE
#> 2022-04-11 15:02:27 - r-draw_manhattanPlot(): Get data ...
#> 2022-04-11 15:02:27 - r-readGWAS(): Read result file ... 
#> 2022-04-11 15:02:27 - r-readGWAS(): Read result file DONE 
#> 2022-04-11 15:02:27 - r-readGWAS(): Convert Json to data.frame ... 
#> 2022-04-11 15:02:27 - r-readGWAS(): Convert Json to data.frame DONE 
#> 2022-04-11 15:02:27 - r-readGWAS(): DONE, return output.
#> 2022-04-11 15:02:27 - r-draw_manhattanPlot(): Get data DONE
#> 2022-04-11 15:02:27 - r-draw_manhattanPlot(): Draw Manhattan Plot ...
#> 2022-04-11 15:02:27 - r-manPlot(): Check parameters...
#> 2022-04-11 15:02:27 - r-manPlot(): Check parameters DONE
#> 2022-04-11 15:02:27 - r-manPlot(): Check chromosome name ...
#> 2022-04-11 15:02:27 - r-manPlot(): Check chromosome name DONE
#> 2022-04-11 15:02:27 - r-manPlot(): Remove NAs ...
#> 2022-04-11 15:02:27 - r-manPlot(): Remove NAs DONE
#> 2022-04-11 15:02:27 - r-manPlot(): Adjust p-values ...
#> 2022-04-11 15:02:27 - r-adjustPval(): Check adj_method ...
#> 2022-04-11 15:02:27 - r-adjustPval(): Check adj_method DONE
#> 2022-04-11 15:02:27 - r-adjustPval(): Check p values ...
#> 2022-04-11 15:02:27 - r-adjustPval(): Check p values DONE
#> 2022-04-11 15:02:27 - r-adjustPval(): Adjust p-values ...
#> 2022-04-11 15:02:27 - r-adjustPval(): Adjust p-values DONE
#> 2022-04-11 15:02:27 - r-adjustPval(): Adjust threshold ...
#> 2022-04-11 15:02:27 - r-adjustPval(): Adjust threshold DONE
#> 2022-04-11 15:02:27 - r-adjustPval(): DONE, return output
#> 2022-04-11 15:02:27 - r-manPlot(): Adjust p-values DONE
#> 2022-04-11 15:02:27 - r-manPlot(): Check duplicated SNP ID ...
#> 2022-04-11 15:02:27 - r-manPlot(): Check duplicated SNP ID DONE
#> 2022-04-11 15:02:27 - r-manPlot(): Extract significant SNP ...
#> 2022-04-11 15:02:27 - r-manPlot(): Extract significant SNP DONE
#> 2022-04-11 15:02:27 - r-filterGWAS(): Check parameters...
#> 2022-04-11 15:02:27 - r-filterGWAS(): Check parameters DONE
#> 2022-04-11 15:02:27 - r-filterGWAS(): Filter points ...
#> 2022-04-11 15:02:27 - r-filterGWAS(): skip filter_pAdj
#> 2022-04-11 15:02:27 - r-filterGWAS(): skip filter_nPoints
#> 2022-04-11 15:02:27 - r-manPlot(): Draw plot ...
#> 2022-04-11 15:02:28 - r-manPlot(): Draw plot DONE
#> 2022-04-11 15:02:28 - r-manPlot(): DONE, return output
#> 2022-04-11 15:02:28 - r-draw_manhattanPlot(): Draw Manhattan Plot DONE
#> 2022-04-11 15:02:28 - r-draw_manhattanPlot(): Save results ...
#> 2022-04-11 15:02:30 - r-draw_manhattanPlot(): Save results DONE
```

``` r
gwas_adj <- run_resAdjustment(gwasFile = gwas_results$file,
                              gwasUrl = NULL,
                              adj_method = "bonferroni",
                              outFile = tempfile(fileext = ".json"))
#> 2022-04-11 15:02:30 - r-run_resAdjustment(): Get data ...
#> 2022-04-11 15:02:30 - r-readGWAS(): Read result file ... 
#> 2022-04-11 15:02:30 - r-readGWAS(): Read result file DONE 
#> 2022-04-11 15:02:30 - r-readGWAS(): Convert Json to data.frame ... 
#> 2022-04-11 15:02:31 - r-readGWAS(): Convert Json to data.frame DONE 
#> 2022-04-11 15:02:31 - r-readGWAS(): DONE, return output.
#> 2022-04-11 15:02:31 - r-run_resAdjustment(): Get data DONE
#> 2022-04-11 15:02:31 - r-run_resAdjustment(): Adjust p-values ...
#> 2022-04-11 15:02:31 - r-adjustPval(): Check adj_method ...
#> 2022-04-11 15:02:31 - r-adjustPval(): Check adj_method DONE
#> 2022-04-11 15:02:31 - r-adjustPval(): Check p values ...
#> 2022-04-11 15:02:31 - r-adjustPval(): Check p values DONE
#> 2022-04-11 15:02:31 - r-adjustPval(): Adjust p-values ...
#> 2022-04-11 15:02:31 - r-adjustPval(): Adjust p-values DONE
#> 2022-04-11 15:02:31 - r-adjustPval(): DONE, return output
#> 2022-04-11 15:02:31 - r-run_resAdjustment(): Adjust p-values DONE
#> 2022-04-11 15:02:31 - r-filterGWAS(): Check parameters...
#> 2022-04-11 15:02:31 - r-filterGWAS(): Check parameters DONE
#> 2022-04-11 15:02:31 - r-filterGWAS(): Filter points ...
#> 2022-04-11 15:02:31 - r-filterGWAS(): skip filter_pAdj
#> 2022-04-11 15:02:31 - r-filterGWAS(): skip filter_quant
#> 2022-04-11 15:02:31 - r-filterGWAS(): skip filter_nPoints
#> 2022-04-11 15:02:31 - r-run_resAdjustment(): Save results ...
#> 2022-04-11 15:02:31 - r-saveGWAS(): Check file ...
#> 2022-04-11 15:02:31 - r-saveGWAS(): Check file DONE
#> 2022-04-11 15:02:31 - r-run_resAdjustment(): Save results DONE
substr(gwas_adj$gwasAdjusted, start=1, stop=500)
#> [
#>   {
#>     "chr": "1",
#>     "pos": 9563,
#>     "id": "SNP-1.8562.",
#>     "A1": "A",
#>     "A2": "T",
#>     "freqA2": 0.1359,
#>     "score": 4.395,
#>     "p": 0.036,
#>     "p_adj": 1
#>   },
#>   {
#>     "chr": "1",
#>     "pos": 25922,
#>     "id": "SNP-1.24921.",
#>     "A1": "C",
#>     "A2": "T",
#>     "freqA2": 0.1254,
#>     "score": 1.6744,
#>     "p": 0.1957,
#>     "p_adj": 1
#>   },
#>   {
#>     "chr": "1",
#>     "pos": 26254,
#>     "id": "SNP-1.25253.",
#>     "A1": "A",
#>     "A2": "T",
#>     "freqA2": 0.2935,
#>     "score": 3.6592,
#>     "p": 0.0558,
#> 
```

</details>

## LD plot

<details>
<summary>
Click to expand
</summary>

This help to calculate and visualize the linkage disequilibrium between
some consecutive snps.

``` sh
docker run --rm -v "$PWD"/data/geno/:/geno \
    -v "$PWD"/readmeTemp:/out rgenotoolsengine \
    ldplot \
    --genoFile "/geno/testMarkerData01.vcf.gz" \
    --from 42 \
    --to 62 \
    --outFile "/out/ldplot.png"
```

**Main function**

``` r
imgFile <- draw_ldPlot(genoFile = "data/geno/testMarkerData01.vcf.gz",
                       genoUrl = NULL,
                       from = 42,
                       to = 62,
                       outFile = tempfile(fileext = ".png")) 
#> 2022-04-11 15:02:31 - r-draw_ldPlot(): Get data ...
#> 2022-04-11 15:02:31 - r-readGenoData(): Check file extention ... 
#> 2022-04-11 15:02:31 - r-readGenoData(): Read geno file ... 
#> ped stats and snps stats have been set. 
#> 'p' has been set. 
#> 'mu' and 'sigma' have been set.
#> 2022-04-11 15:02:32 - r-readGenoData(): Read geno file DONE 
#> 2022-04-11 15:02:32 - r-readGenoData(): DONE, return output.
#> 2022-04-11 15:02:32 - r-draw_ldPlot(): Get data DONE
#> 2022-04-11 15:02:32 - r-draw_ldPlot(): Draw LD Plot ...
#> 2022-04-11 15:02:32 - r-LDplot(): Check "from" and "to" format ...
#> 2022-04-11 15:02:32 - r-LDplot(): Check "from" and "to" format DONE
#> 2022-04-11 15:02:32 - r-LDplot(): Check "from" < "to"...
#> 2022-04-11 15:02:32 - r-LDplot(): Check "from" < "to" DONE
#> 2022-04-11 15:02:32 - r-LDplot(): Check number of SNP < 50...
#> 2022-04-11 15:02:32 - r-LDplot(): Check number of SNP < 50 DONE
#> 2022-04-11 15:02:32 - r-LDplot(): Check file ...
#> 2022-04-11 15:02:32 - r-LDplot(): Check file DONE
#> 2022-04-11 15:02:32 - r-LDplot(): Compute LD ...
#> 2022-04-11 15:02:32 - r-LDplot(): Compute LD DONE
#> 2022-04-11 15:02:32 - r-LDplot(): Create LD plot ...
#> 2022-04-11 15:02:32 - r-LDplot(): Create create file: /tmp/Rtmpi6YPG5/file1e198a3e768f9b.png
#> 2022-04-11 15:02:32 - r-LDplot(): Create LD plot DONE
#> 2022-04-11 15:02:32 - r-LDplot(): DONE, return output
#> 2022-04-11 15:02:32 - r-draw_ldPlot(): Draw LD Plot DONE
```

</details>

## Relationship matrix

<details>
<summary>
Click to expand
</summary>

Relationship matrix represents how close two individuals are to each
other (share the same genes). It can be calculated using their pedigree,
or their genotypes.

### Pedigree relationship matrix

This engine can calculate the pedigree-based relationship matrix:

``` sh
docker run --rm -v "$PWD"/data/pedigree/:/pedigree \
    -v "$PWD"/readmeTemp:/out rgenotoolsengine \
    relmat-ped \
    --pedFile "/pedigree/testPedData_char.csv" \
    --outFile "/out/pedRelMat.json"
```

``` r
calc_pedRelMAt(pedFile = 'data/pedigree/testPedData_char.csv',
               pedUrl = NULL,
               header = TRUE,
               unknown_string = '',
               outFile = tempfile(fileext = ".json"))
#> 2022-04-11 15:02:32 - r-calc_pedRelMAt(): Get data ...
#> 2022-04-11 15:02:32 - r-readPedData: Read pedigree file ...
#> 2022-04-11 15:02:32 - r-readPedData: Read pedigree file DONE
#> 2022-04-11 15:02:32 - r-readPedData: Check pedigree file ...
#> 2022-04-11 15:02:32 - r-readPedData: Check pedigree file DONE
#> 2022-04-11 15:02:32 - r-readPedData: DONE, return output.
#> 2022-04-11 15:02:32 - r-calc_pedRelMAt(): Get data DONE
#> 2022-04-11 15:02:32 - r-calc_pedRelMAt(): Calcualte pedigree relationship matrix ...
#> 2022-04-11 15:02:32 - r-pedRelMat(): Check inputs ...
#> 2022-04-11 15:02:32 - r-pedRelMat(): Check inputs DONE
#> 2022-04-11 15:02:32 - r-pedRelMat(): Create look-up table ...
#> 2022-04-11 15:02:32 - r-pedRelMat(): Create look-up table DONE
#> 2022-04-11 15:02:32 - r-pedRelMat(): Calculate relationship matrix ...
#> 2022-04-11 15:02:32 - r-pedRelMat(): Calculate relationship matrix DONE
#> 2022-04-11 15:02:32 - r-pedRelMat(): DONE, return output.
#> 2022-04-11 15:02:32 - r-calc_pedRelMAt(): Calcualte pedigree relationship matrix DONE
#> 2022-04-11 15:02:32 - r-calc_pedRelMAt(): Get metadata ...
#> 2022-04-11 15:02:32 - r-calc_pedRelMAt(): Get metadata DONE
#> 2022-04-11 15:02:32 - r-calc_pedRelMAt(): Save results ...
#> 2022-04-11 15:02:32 - r-saveRelMat(): Check relationship matrix ...
#> 2022-04-11 15:02:32 - r-saveRelMat(): Check relationship matrix DONE
#> 2022-04-11 15:02:32 - r-saveRelMat(): Check file ...
#> 2022-04-11 15:02:32 - r-saveRelMat(): Check file DONE
#> 2022-04-11 15:02:32 - r-saveRelMat(): Check file format ...
#> 2022-04-11 15:02:32 - r-saveRelMat(): Check file format DONE
#> 2022-04-11 15:02:32 - r-saveRelMat(): Write relationship matrix in `.json` file ...
#> 2022-04-11 15:02:32 - r-saveRelMat(): Write relationship matrix in `.json` file DONE
#> 2022-04-11 15:02:32 - r-calc_pedRelMAt(): Save results DONE
#> $relMat
#>               Pluto     Zeus     Leda Dione   Tantale   Europe     Pelos  Minos
#> Pluto        1.0000 0.000000 0.000000 0.000 0.5000000 0.000000 0.2500000 0.0000
#> Zeus         0.0000 1.000000 0.000000 0.000 0.5000000 0.500000 0.2500000 0.7500
#> Leda         0.0000 0.000000 1.000000 0.000 0.0000000 0.500000 0.0000000 0.2500
#> Dione        0.0000 0.000000 0.000000 1.000 0.0000000 0.000000 0.5000000 0.0000
#> Tantale      0.5000 0.500000 0.000000 0.000 1.0000000 0.250000 0.5000000 0.3750
#> Europe       0.0000 0.500000 0.500000 0.000 0.2500000 1.000000 0.1250000 0.7500
#> Pelos        0.2500 0.250000 0.000000 0.500 0.5000000 0.125000 1.0000000 0.1875
#> Minos        0.0000 0.750000 0.250000 0.000 0.3750000 0.750000 0.1875000 1.2500
#> Pelopia      0.2500 0.250000 0.000000 0.500 0.5000000 0.125000 1.0000000 0.1875
#> Atree        0.2500 0.250000 0.000000 0.500 0.5000000 0.125000 1.0000000 0.1875
#> Catree       0.0000 0.375000 0.125000 0.000 0.1875000 0.375000 0.0937500 0.6250
#> Egiste       0.2500 0.250000 0.000000 0.500 0.5000000 0.125000 1.0000000 0.1875
#> Aerope       0.0000 0.187500 0.062500 0.000 0.0937500 0.187500 0.0468750 0.3125
#> Menelas      0.1250 0.218750 0.031250 0.250 0.2968750 0.156250 0.5234375 0.2500
#> Agamemnon    0.1250 0.218750 0.031250 0.250 0.2968750 0.156250 0.5234375 0.2500
#> Clytemnestre 0.0000 0.500000 0.500000 0.000 0.2500000 0.500000 0.1250000 0.5000
#> Helene       0.0000 0.500000 0.500000 0.000 0.2500000 0.500000 0.1250000 0.5000
#> Electre      0.0625 0.359375 0.265625 0.125 0.2734375 0.328125 0.3242188 0.3750
#> Oreste       0.0625 0.359375 0.265625 0.125 0.2734375 0.328125 0.3242188 0.3750
#> Hyphigenie   0.0625 0.359375 0.265625 0.125 0.2734375 0.328125 0.3242188 0.3750
#>                Pelopia     Atree    Catree    Egiste    Aerope   Menelas
#> Pluto        0.2500000 0.2500000 0.0000000 0.2500000 0.0000000 0.1250000
#> Zeus         0.2500000 0.2500000 0.3750000 0.2500000 0.1875000 0.2187500
#> Leda         0.0000000 0.0000000 0.1250000 0.0000000 0.0625000 0.0312500
#> Dione        0.5000000 0.5000000 0.0000000 0.5000000 0.0000000 0.2500000
#> Tantale      0.5000000 0.5000000 0.1875000 0.5000000 0.0937500 0.2968750
#> Europe       0.1250000 0.1250000 0.3750000 0.1250000 0.1875000 0.1562500
#> Pelos        1.0000000 1.0000000 0.0937500 1.0000000 0.0468750 0.5234375
#> Minos        0.1875000 0.1875000 0.6250000 0.1875000 0.3125000 0.2500000
#> Pelopia      1.5000000 1.0000000 0.0937500 1.2500000 0.0468750 0.5234375
#> Atree        1.0000000 1.5000000 0.0937500 1.0000000 0.0468750 0.7734375
#> Catree       0.0937500 0.0937500 1.0000000 0.0937500 0.5000000 0.2968750
#> Egiste       1.2500000 1.0000000 0.0937500 1.5000000 0.0468750 0.5234375
#> Aerope       0.0468750 0.0468750 0.5000000 0.0468750 1.0000000 0.5234375
#> Menelas      0.5234375 0.7734375 0.2968750 0.5234375 0.5234375 1.0234375
#> Agamemnon    0.5234375 0.7734375 0.2968750 0.5234375 0.5234375 0.6484375
#> Clytemnestre 0.1250000 0.1250000 0.2500000 0.1250000 0.1250000 0.1250000
#> Helene       0.1250000 0.1250000 0.2500000 0.1250000 0.1250000 0.1250000
#> Electre      0.3242188 0.4492188 0.2734375 0.3242188 0.3242188 0.3867188
#> Oreste       0.3242188 0.4492188 0.2734375 0.3242188 0.3242188 0.3867188
#> Hyphigenie   0.3242188 0.4492188 0.2734375 0.3242188 0.3242188 0.3867188
#>              Agamemnon Clytemnestre Helene   Electre    Oreste Hyphigenie
#> Pluto        0.1250000       0.0000 0.0000 0.0625000 0.0625000  0.0625000
#> Zeus         0.2187500       0.5000 0.5000 0.3593750 0.3593750  0.3593750
#> Leda         0.0312500       0.5000 0.5000 0.2656250 0.2656250  0.2656250
#> Dione        0.2500000       0.0000 0.0000 0.1250000 0.1250000  0.1250000
#> Tantale      0.2968750       0.2500 0.2500 0.2734375 0.2734375  0.2734375
#> Europe       0.1562500       0.5000 0.5000 0.3281250 0.3281250  0.3281250
#> Pelos        0.5234375       0.1250 0.1250 0.3242188 0.3242188  0.3242188
#> Minos        0.2500000       0.5000 0.5000 0.3750000 0.3750000  0.3750000
#> Pelopia      0.5234375       0.1250 0.1250 0.3242188 0.3242188  0.3242188
#> Atree        0.7734375       0.1250 0.1250 0.4492188 0.4492188  0.4492188
#> Catree       0.2968750       0.2500 0.2500 0.2734375 0.2734375  0.2734375
#> Egiste       0.5234375       0.1250 0.1250 0.3242188 0.3242188  0.3242188
#> Aerope       0.5234375       0.1250 0.1250 0.3242188 0.3242188  0.3242188
#> Menelas      0.6484375       0.1250 0.1250 0.3867188 0.3867188  0.3867188
#> Agamemnon    1.0234375       0.1250 0.1250 0.5742188 0.5742188  0.5742188
#> Clytemnestre 0.1250000       1.0000 0.5000 0.5625000 0.5625000  0.5625000
#> Helene       0.1250000       0.5000 1.0000 0.3125000 0.3125000  0.3125000
#> Electre      0.5742188       0.5625 0.3125 1.0625000 0.5683594  0.5683594
#> Oreste       0.5742188       0.5625 0.3125 0.5683594 1.0625000  0.5683594
#> Hyphigenie   0.5742188       0.5625 0.3125 0.5683594 0.5683594  1.0625000
#> 
#> $metadata
#> $metadata$info
#> [1] "R-geno-engine, Pedigree relationship matrix"
#> 
#> $metadata$date
#> [1] "2022-04-11 15:02:32 JST"
#> 
#> $metadata$nInds
#> [1] 20
#> 
#> $metadata$pedFP
#> [1] "b15d79357d4b632fce0d63e9444d9ce3"
#> 
#> 
#> $file
#> [1] "/tmp/Rtmpi6YPG5/file1e198a4a459ab1.json"
```

### Relationship matrix visualisation

The engine include a function to generate an heatmap of a relationship
matrix:

This heatmap can either be interactive (`.html` file) or static (`.png`
file).

![heatmap](./data/results/relationshipHeatmap.png)

``` sh
docker run --rm -v "$PWD"/data/results/:/results \
    -v "$PWD"/readmeTemp:/out rgenotoolsengine \
    relmat-heatmap \
    --relmatFile "/results/pedRelMat.json" \
    --outFile "/out/relMat_heatmap.png"
```

``` r
draw_relHeatmap(relMatFile = 'data/results/pedigreeRelationship.csv',
                relMatUrl = NULL,
                interactive = FALSE,
                outFile = tempfile(fileext = ".png"))
#> 2022-04-11 15:02:32 - r-draw_relHeatmap(): Check outFile ...
#> 2022-04-11 15:02:32 - r-draw_relHeatmap(): Check outFile DONE
#> 2022-04-11 15:02:32 - r-draw_relHeatmap(): Get data ...
#> 2022-04-11 15:02:32 - r-readRelMat(): Check file format ...
#> 2022-04-11 15:02:32 - r-readRelMat(): Check file format DONE
#> 2022-04-11 15:02:32 - r-readRelMat(): Read relationship matrix `csv` file ... 
#> 2022-04-11 15:02:32 - r-readRelMat(): Read relationship matrix `csv` file DONE
#> 2022-04-11 15:02:32 - r-readRelMat(): Check loaded relationship matrix ...
#> 2022-04-11 15:02:32 - r-readRelMat(): Check loaded relationship matrix DONE
#> 2022-04-11 15:02:32 - r-readRelMat(): DONE, return output.
#> 2022-04-11 15:02:32 - r-draw_relHeatmap(): Get data DONE
#> 2022-04-11 15:02:32 - r-draw_relHeatmap(): Open connexion to draw the png plot ...
#> 2022-04-11 15:02:32 - r-draw_relHeatmap(): Open connexion to draw the png plot DONE
#> 2022-04-11 15:02:32 - r-draw_relHeatmap(): Draw relationship heatmap ...
#> 2022-04-11 15:02:32 - r-manPlot(): Check parameters ...
#> 2022-04-11 15:02:32 - r-manPlot(): Check parameters DONE
#> 2022-04-11 15:02:32 - r-manPlot(): Create static heatmap ...
#> 2022-04-11 15:02:32 - r-manPlot(): Create static heatmap DONE
#> 2022-04-11 15:02:32 - r-manPlot(): DONE, return output
#> 2022-04-11 15:02:32 - r-draw_relHeatmap(): Draw relationship heatmap DONE
#> 2022-04-11 15:02:32 - r-draw_relHeatmap(): Save results ...
#> 2022-04-11 15:02:32 - r-draw_relHeatmap(): Save results DONE
#> NULL
```

</details>

## Pedigree network

<details>
<summary>
Click to expand
</summary>

***This is quite experimental***

The engine contain a function to generate an interactive network of the
pedigree relations.

Individuals with parental relations are linked with an arrow pointing
from the parents to the offspring.

![pedNet](README_files/pedNet.png)

> Note: This network is not organized by generation (like a genealogical
> tree) can be big, messy and not very responsive if the number of
> individual in the pedigree is big.

``` sh
docker run --rm -v "$PWD"/data/pedigree/:/pedigree \
    -v "$PWD"/readmeTemp:/out rgenotoolsengine \
    pedNetwork \
    --pedFile "/pedigree/testPedData_char.csv" \
    --outFile "/out/pedNet.html"
```

**Main function**

``` r
imgFile <- draw_pedNetwork(pedFile = "data/pedigree/testPedData_char.csv",
                           pedUrl = NULL,
                           unknown_string = '',
                           header = TRUE,
                           outFile = tempfile(fileext = ".html")) 
#> 2022-04-11 15:02:32 - r-draw_pedNetwork(): Get data ...
#> 2022-04-11 15:02:32 - r-readPedData: Read pedigree file ...
#> 2022-04-11 15:02:32 - r-readPedData: Read pedigree file DONE
#> 2022-04-11 15:02:32 - r-readPedData: Check pedigree file ...
#> 2022-04-11 15:02:32 - r-readPedData: Check pedigree file DONE
#> 2022-04-11 15:02:32 - r-readPedData: DONE, return output.
#> 2022-04-11 15:02:32 - r-draw_pedNetwork(): Get data DONE
#> 2022-04-11 15:02:32 - r-draw_pedNetwork(): Draw pedigree interactive network ...
#> 2022-04-11 15:02:32 - r-pedNetwork(): Check parameters ...
#> 2022-04-11 15:02:32 - r-pedNetwork(): Check inputs ...
#> 2022-04-11 15:02:32 - r-pedNetwork(): Check inputs DONE
#> 2022-04-11 15:02:32 - r-pedNetwork(): Create network data ...
#> 2022-04-11 15:02:32 - r-pedNetwork(): Create network data DONE
#> 2022-04-11 15:02:32 - r-pedNetwork(): Create network ...
#> 2022-04-11 15:02:32 - r-pedNetwork(): Create network DONE
#> 2022-04-11 15:02:32 - r-pedNetwork(): DONE, return output.
#> 2022-04-11 15:02:32 - r-draw_pedNetwork(): Draw pedigree interactive network DONE
#> 2022-04-11 15:02:32 - r-draw_pedNetwork(): Save results ...
#> 2022-04-11 15:02:33 - r-draw_pedNetwork(): Save results DONE
```

</details>

# Data references

Some example data are available in the folder [data](data/). This folder
also includes examples of output files.

These examples output files can be automatically generated using the
function `createResultExample` (defined in
[`./src/utils.R`](src/utils.R)):

``` r
createResultExample()
```

The genotypic and phenotypic data used as example come from:

> Keyan Zhao, Chih-Wei Tung, Georgia C. Eizenga, Mark H. Wright, M.
> Liakat Ali, Adam H. Price, Gareth J. Norton, M. Rafiqul Islam, Andy
> Reynolds, Jason Mezey, Anna M. McClung, Carlos D. Bustamante & Susan
> R. McCouch (2011). [Genome-wide association mapping reveals a rich
> genetic architecture of complex traits in *Oryza
> sativa*.](http://www.nature.com/ncomms/journal/v2/n9/full/ncomms1467.html)
> Nat Comm 2:467 \| DOI: 10.1038/ncomms1467, Published Online 13 Sep
> 2011.

> Flutre, T., Diot, J., and David, J. (2019). [PlantBreedGame: A Serious
> Game that Puts Students in the Breeder’s Seat. Crop
> Science.](https://acsess.onlinelibrary.wiley.com/doi/10.2135/cropsci2019.03.0183le)
> DOI 10.2135/cropsci2019.03.0183le

# Utils

The [`utils`](./utils) folder contain miscellaneous code that can be
useful for developers working on this project. This engine should NOT
depend on any file of this folder.
