---
output:
  github_document:
    toc: true
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
dir.create("readmeTemp")
```

# GWAS-Engine

<!-- badges: start -->
<!-- badges: end -->

The goal of GWAS-Engine is to provide simple "input/output" style code for GWAS analysis in order to be used inside an API or other services.


# Introduction

GWAS stands for [Genome-Wide Association Study](https://en.wikipedia.org/wiki/Genome-wide_association_study). It is a statistical study that aims to detect genetic markers associated with a particular phenotypic trait.

To test if a marker is associated with a particular phenotypic trait, we split individuals into different groups according to their genotype for the studied marker and we test if there is a statistical difference between the phenotypic values of the groups. The strength of the statistical difference is measured by what statisticians call the "[p-value](https://en.wikipedia.org/wiki/P-value)".

Usually to consider a test "statistically significant" we check if the p-value of the test is lower than a threshold "α". In this case the test is considered positive (*"there is a statistical difference"*) but we have a probability α to have a false positive. Very often α is equal to 5% or 1%.

GWAS analysis does this test individually and independently for all the genetic markers in the data-set and records the "statistical significance" (the p-values) of all the tests.

However, the number of tested genetic markers are usually very large (several hundreds of thousands or several million) therefore, the number of false-positive can be high even for a small value of α. For example, with α=1%, 500000 markers, and if none of them is associated with the phenotypic trait, we can still expect 5000 "significant test".

To avoid that, the p-values must be [adjusted](https://en.wikipedia.org/wiki/Multiple_comparisons_problem) according to the number of markers.

Finally, the result of a GWAS analysis is presented as a [Manhattan_plot](https://en.wikipedia.org/wiki/Manhattan_plot) which shows the `-log(p-value)` on the y-axis and all the markers (order according to their position on the chromosomes) on the x-axis. A horizontal line represents the significance threshold and the points of the markers above the line can be considered as being associated with the phenotypic trait.

![screenshot of the plotly graph](README_files/manPlot.png)


GWAS-Engine provides command line tools to do GWAS analysis.

> Note: You can also check this [7min video](https://www.youtube.com/watch?v=sOP8WacfBM8) by Nuno Carvalho that explain GWAS very well. 


# Build GWAS-Engine

The easiest way to use GWAS-Engine is to use the [Docker](https://www.docker.com/) image. To build the docker image simply do:

```sh
git clone https://github.com/ut-biomet/GWAS-Engine
cd GWAS-Engine
docker build -t gwasengine ./
```

## Test the image

You can test that the image is build correctly by running:

```sh
docker run --entrypoint="Rscript" gwasengine ./tests/testthat.R
```

All tests should pass.

# How to use GWAS-Engine

Basically the command to use the engine from the docker image is as follow:

```sh
docker run gwasengine tool --param1 valueOfParam1 --param2 valueOfParam2 ...
```

- `tool` can be either:
    - `gwas`: to do a gwas analysis from geno and pheno files
    - `manplot`: to create an interactive manhathan plot from the results generated by `gwas`
    - `adjresults`: to calculate the adjusted p-value from the results generated by `gwas`
    - `ldplot`: to get a plot showing the Linkage disequilibrium between consecutive markers.

The other parameters depends of the tool and you can have the exhaustive list of them with the command:

```sh
docker run gwasengine
```

<details><summary>Click to see outputs</summary>

```{sh}
docker run gwasengine
```

</details>


However, in order to let the docker container read/write files from the host machine, you need to mount the the folders containing the files in the container using the option `-v --volume`. Then the file path must be specify relative to the place in the container. (See examples below).

## Examples

```sh
docker run -v "$PWD"/data/geno/:/geno \
    -v "$PWD"/data/pheno/:/pheno \
    -v "$PWD"/readmeTemp:/out gwasengine \
    gwas \
    --genoFile "/geno/testMarkerData01.vcf.gz" \
    --phenoFile "/pheno/testPhenoData01.csv" \
    --trait "Flowering.time.at.Arkansas" \
    --test "score" \
    --fixed 0 \
    --response "quantitative" \
    --thresh_maf 0.05 \
    --thresh_callrate 0.95 \
    --outFile "/out/gwasRes.json"
```

<details><summary>Click to see outputs</summary>

```{sh}
docker run -v "$PWD"/data/geno/:/geno \
    -v "$PWD"/data/pheno/:/pheno \
    -v "$PWD"/readmeTemp:/out gwasengine \
    gwas \
    --genoFile "/geno/testMarkerData01.vcf.gz" \
    --phenoFile "/pheno/testPhenoData01.csv" \
    --trait "Flowering.time.at.Arkansas" \
    --test "score" \
    --fixed 0 \
    --response "quantitative" \
    --thresh_maf 0.05 \
    --thresh_callrate 0.95 \
    --outFile "/out/gwasRes.json"
```

</details>

### Draw Manhattan Plot


```sh
docker run -v "$PWD"/readmeTemp:/files gwasengine \
    manplot \
    --gwasFile "/files/gwasRes.json" \
    --adj_method "bonferroni" \
    --thresh_p 0.05 \
    --interactive TRUE \
    --outFile "/files/manPlot.html"
```

<details><summary>Click to see outputs</summary>

**Interactive plot**

```{sh}
docker run -v "$PWD"/readmeTemp:/files gwasengine \
    manplot \
    --gwasFile "/files/gwasRes.json" \
    --adj_method "bonferroni" \
    --thresh_p 0.05 \
    --interactive TRUE \
    --outFile "/files/manPlot.html"
```

![screenshot of the plotly graph](README_files/manPlot.png)

**Static plot**

```{sh}
docker run -v "$PWD"/readmeTemp:/files gwasengine \
    manplot \
    --gwasFile "/files/gwasRes.json" \
    --adj_method "bonferroni" \
    --thresh_p 0.05 \
    --interactive FALSE \
    --outFile "/files/manPlot.png"
```


```{r include=FALSE}
file.copy('readmeTemp/manPlot.png', 'README_files/manPlotStatic_1.png', overwrite = TRUE)
```


```{r echo=FALSE}
knitr::include_graphics('README_files/manPlotStatic_1.png')
```


</details>

### Adjust p-values


```sh
docker run -v "$PWD"/readmeTemp:/files gwasengine \
    adjresults \
    --gwasFile "/files/gwasRes.json" \
    --adj_method "bonferroni" \
    --outFile "/files/adjRes.json"
```

<details><summary>Click to see outputs</summary>

```{sh}
docker run -v "$PWD"/readmeTemp:/files gwasengine \
    adjresults \
    --gwasFile "/files/gwasRes.json" \
    --adj_method "bonferroni" \
    --outFile "/files/adjRes.json"
```


</details>


### Draw LD plot


```sh
docker run -v "$PWD"/data/geno/:/geno \
    -v "$PWD"/readmeTemp:/out gwasengine \
    ldplot \
    --genoFile "/geno/testMarkerData01.vcf.gz" \
    --from 42 \
    --to 62 \
    --outFile "/out/ldplot.png"
```

<details><summary>Click to see outputs</summary>

```{sh}
docker run -v "$PWD"/data/geno/:/geno \
    -v "$PWD"/readmeTemp:/out gwasengine \
    ldplot \
    --genoFile "/geno/testMarkerData01.vcf.gz" \
    --from 42 \
    --to 62 \
    --outFile "/out/ldplot.png"
```

```{r}
library(png)
img <- readPNG("readmeTemp/ldplot.png")
grid::grid.raster(img)
```

</details>

# For `R` developpers:

This GWAS-Engine is written in `R`, therefore it can also be run directly from your computer (without docker) if you have all the dependencies installed. To do so you can execute the `gwas-engine.R` file with the `Rscript` command.

For example from a bash terminal:

```sh
Rscript ./gwas-engine.R \
    gwas \
    --genoFile "data/geno/testMarkerData01.vcf.gz" \
    --phenoFile "data/pheno/testPhenoData01.csv" \
    --trait "Flowering.time.at.Arkansas" \
    --test "score" \
    --fixed 0 \
    --response "quantitative" \
    --thresh_maf 0.05 \
    --thresh_callrate 0.95 \
    --outFile "readmeTemp/gwasRes.json"
```


```sh
Rscript ./gwas-engine.R \
    manplot \
    --gwasFile "data/results/gwasResult.json" \
    --adj_method "bonferroni" \
    --thresh_p 0.05 \
    --interactive FALSE \
    --outFile "readmeTemp/manplot.png"
```



## Dependencies

Package dependencies are managed using [Renv](https://rstudio.github.io/renv/articles/renv.html). To install dependencies do in a R console: `renv::restore()`. (you need `/path/to/GWAS-Engine` as working directory).


```{r dep, results='hide'}
# required packages for using GWAS-Engine:
library(argparser) # Command-Line Argument Parser
library(gaston) # for many functions
library(jsonlite) # manage json format
library(manhattanly) # manhattan plot using plotly
library(R6) # R object oriented
```

## GWAS-Engine as a R library

It is possible to use GWAS-Engine as a library to be used in other R projects.

To do so, you can load all the code of this engine using 

```r
sapply(FUN = source,
       X = list.files("/path/to/GWAS-Engine/src", pattern = ".R$", full.names = T))
```

```{r source, include=FALSE}
invisible(
  sapply(FUN = source,
         X = list.files("src", pattern = ".R$",full.names = T))
)
```


### Main functions

GWAS-Engine provide 4 main R functions similar to what can be done with command line arguments.

They take as inputs file paths and other specific argument and write the results in output files:

```{r runGWAS}
gwas_results <- run_gwas(genoFile = "data/geno/testMarkerData01.vcf.gz",
                         phenoFile = "data/pheno/testPhenoData01.csv",
                         genoUrl = NULL,
                         phenoUrl = NULL,
                         trait = "Flowering.time.at.Arkansas",
                         test = "score",
                         fixed = 0,
                         response = "quantitative",
                         thresh_maf = 0.05,
                         thresh_callrate = 0.95,
                         outFile = tempfile(fileext = ".json"))
gwas_results$file
substr(gwas_results$gwasRes, start=1, stop=500)
```

```{r draw_manhattanPlot}
p <- draw_manhattanPlot(gwasFile = gwas_results$file,
                        gwasUrl = NULL,
                        adj_method = "bonferroni",
                        thresh_p = 0.05,
                        chr = NA,
                        interactive = TRUE,
                        outFile = tempfile(fileext = ".html"))
```
```{r eval=FALSE, include=FALSE}
p
```



```{r draw_manhattanPlot2}
manPlotFile <- tempfile(fileext = ".png")
p <- draw_manhattanPlot(gwasFile = gwas_results$file,
                        gwasUrl = NULL,
                        adj_method = "bonferroni",
                        thresh_p = 0.05,
                        chr = NA,
                        interactive = FALSE,
                        outFile = manPlotFile)
```

```{r include=FALSE}
file.copy(manPlotFile, 'README_files/manPlotStatic_2.png', overwrite = TRUE)
```

```{r echo=FALSE}
knitr::include_graphics('README_files/manPlotStatic_2.png')
```


```{r adj_pvals}
gwas_adj <- run_resAdjustment(gwasFile = gwas_results$file,
                              gwasUrl = NULL,
                              adj_method = "bonferroni",
                              outFile = tempfile(fileext = ".json"))
substr(gwas_adj$gwasAdjusted, start=1, stop=500)
```

```{r draw_ldPlot}
imgFile <- draw_ldPlot(genoFile = "data/geno/testMarkerData01.vcf.gz",
                       genoUrl = NULL,
                       from = 42,
                       to = 62,
                       outFile = tempfile(fileext = ".png")) 
```

> See: [src/mainFunctions.R](src/mainFunctions.R)

### Other functions

Other R functions used by the engine are defined in files:

- [src/gwas.R](src/gwas.R)
- [src/readWriteData.R](src/readWriteData.R)
- [src/plot.R](src/plot.R)

For example:

#### Load data

Example data are stored in the `data` folder.

`readData` read and prepare the data for GWAS analysis
```{r loadDta, eval=FALSE}
data <- readData(genoFile = "data/geno/testMarkerData01.vcf",
                 phenoFile = "data/pheno/testPhenoData01.csv")
```

For online data, they can be loaded using:

```{r downloadDta, eval=FALSE}
data <- downloadData(genoUrl = "url/to/geno_data.vcf",
                     phenoUrl = "url/to/pheno_data.csv")
```

> See: [src/readWriteData.R](src/readWriteData.R)

#### GWAS

The `gwas` function return a `data.frame`.

```{r, eval=FALSE}
gwasRes <- gwas(data = data,
                trait = "Flowering.time.at.Arkansas",
                test = "score",
                fixed = 0,
                response = "quantitative",
                thresh_maf = 0.05,
                thresh_callrate = 0.95)
head(gwasRes)
```


Save GWAS result in a `.json` file.

```{r, eval=FALSE}
(file <- saveGWAS(gwasRes, metadata = "README"))
cat(paste(readLines(file, n=20), collapse = "\n"))
```

#### Manhattan plot

This function generates a "plotly" graph.

```{r manPlot, eval=FALSE}
p <- manPlot(gwas = gwasRes,
             adj_method = "bonferroni",
             thresh_p = 0.05,
             chr = NA,
             title = "Readme Example")
```
<!-- ![screenshot of the plotly graph](README_files/manPlot2.png) -->

```{r eval=FALSE, include=FALSE}
p
```



#### LD PLot

Compute $r^2$ Linkage Disequilibrium (LD) between given SNPs and save a plot in a temporary PNG file.

```{r, eval=FALSE}
gDta <- readGenoData("data/geno/testMarkerData01.vcf.gz")
(imgFile <- LDplot(geno = gDta,
                   from = 1,
                   to = 20,
                   file = tempfile(fileext = ".png")))
```


```{r, eval=FALSE}
img <- readPNG(imgFile)
grid::grid.raster(img)
```


### Functions documentations

All the functions of this engine use `roxygen2` for documentation. You can generate a markdown documentation located in [`./doc/README.md`](doc/README.md) using the function `writeDoc` (defined in [`./src/utils.R`](src/utils.R)):

```{r eval=FALSE}
writeDoc(srcDir = "src",
         docDir = "doc")
```


# Tests

The R package `testthat` is needed to run the unit tests.
To run the unit tests of this engine you can use the command:

```sh
Rscript ./tests/testthat.R
```


# Data references

Some example data are available in the folder [data](data/). This folder also includes examples of output files.


The genotypic and phenotypic data used as example come from:

> Keyan Zhao, Chih-Wei Tung, Georgia C. Eizenga, Mark H. Wright, M. Liakat Ali, Adam H. Price, Gareth J. Norton, M. Rafiqul Islam, Andy Reynolds, Jason Mezey, Anna M. McClung, Carlos D. Bustamante & Susan R. McCouch (2011). [Genome-wide association mapping reveals a rich genetic architecture of complex traits in *Oryza sativa*.](http://www.nature.com/ncomms/journal/v2/n9/full/ncomms1467.html) Nat Comm 2:467 | DOI: 10.1038/ncomms1467, Published Online 13 Sep 2011.

```{r, include = FALSE}
# remove temp dir
unlink("readmeTemp", recursive = TRUE)
```
