---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# GWAS-Engine

<!-- badges: start -->
<!-- badges: end -->

The goal of GWAS-Engine is to provide simple "input/output" style code for GWAS analysis in order to be used inside an API.


# How to use GWAS-Engine


## Dependencies

```{r dep}
# required packages for using GWAS-Engine:
library(gaston) # for many functions
library(jsonlite) # manage json format
library(manhattanly) # manhattan plot using plotly
library(R6) # R object oriented
```

## Load R scripts

Load all `.R` files in folder `src`.
 
```{r source}
invisible(
  sapply(FUN = source,
         X = list.files("src", pattern = ".R$",full.names = T))
)
```


## Load data

Example data are stored in the `data` folder.

```{r loadDta}
gDta <- readGenoData("data/markers/testMarkerData01.vcf.gz")
pDta <- readPhenoData("data/pheno/testPhenoData01.csv")
```

For online data, they can be loaded using:

```{r downloadDta, eval=FALSE}
gDta <- downloadGenoData("url/to/geno_data.vcf")
pDta <- downloadPhenoData("url/to/pheno_data.csv")
dta <- downloadData("url/to/geno_data.vcf", "url/to/pheno_data.csv")
```


## GWAS

To run a GWAS analysis you should first process the data using `prepareData` and then call the `gwas` function.

The `gwas` function return a `data.frame`.

```{r}
data <- prepareData(gDta=gDta, pDta = pDta)

gwasRes <- gwas(data = data,
                trait = "Panicle.length",
                test = "score",
                fixed = 0,
                response = "quantitative",
                thresh_maf = 0.05,
                thresh_callrate = 0.95)
head(gwasRes)
```


Save GWAS result in a `.json` file.

```{r}
(file <- saveGWAS(gwasRes))
cat(paste(readLines(file, n=20), collapse = "\n"))
```

## Manhattan plot

This function generates a "plotly" graph.

```{r eval=FALSE}
manPlot(gwas = gwasRes,
        adj_method = "bonferroni",
        thresh_p = 0.05,
        chr = NA,
        title = "Readme Example")
```
![screenshot of the plotly graph](README_files/manPlot.png)




## LD PLot

Compute $r^2$ Linkage Disequilibrium (LD) between given SNPs and save a plot in a temporary PNG file.

```{r}
(imgFile <- LDplot(gDta, 1, 20, write = TRUE))
```


```{r}
library(png)
img <- readPNG(imgFile)
grid::grid.raster(img)
```
