
-   [GWAS-Engine](#gwas-engine)
-   [Introduction](#introduction)
-   [Build GWAS-Engine](#build-gwas-engine)
    -   [Test the image](#test-the-image)
-   [How to use GWAS-Engine](#how-to-use-gwas-engine)
    -   [Examples](#examples)
        -   [Draw Manhattan Plot](#draw-manhattan-plot)
        -   [Adjust p-values](#adjust-p-values)
        -   [Draw LD plot](#draw-ld-plot)
-   [For `R` developpers:](#for-r-developpers)
    -   [Dependencies](#dependencies)
    -   [GWAS-Engine as a R library](#gwas-engine-as-a-r-library)
        -   [Main functions](#main-functions)
        -   [Other functions](#other-functions)
        -   [Functions documentations](#functions-documentations)
-   [Tests](#tests)
-   [Data references](#data-references)

<!-- README.md is generated from README.Rmd. Please edit that file -->

# GWAS-Engine

<!-- badges: start -->
<!-- badges: end -->

The goal of GWAS-Engine is to provide simple “input/output” style code
for GWAS analysis in order to be used inside an API or other services.

# Introduction

GWAS stands for [Genome-Wide Association
Study](https://en.wikipedia.org/wiki/Genome-wide_association_study). It
is a statistical study that aims to detect genetic markers associated
with a particular phenotypic trait.

To test if a marker is associated with a particular phenotypic trait, we
split individuals into different groups according to their genotype for
the studied marker and we test if there is a statistical difference
between the phenotypic values of the groups. The strength of the
statistical difference is measured by what statisticians call the
“[p-value](https://en.wikipedia.org/wiki/P-value)”.

Usually to consider a test “statistically significant” we check if the
p-value of the test is lower than a threshold “α”. In this case the test
is considered positive (*“there is a statistical difference”*) but we
have a probability α to have a false positive. Very often α is equal to
5% or 1%.

GWAS analysis does this test individually and independently for all the
genetic markers in the data-set and records the “statistical
significance” (the p-values) of all the tests.

However, the number of tested genetic markers are usually very large
(several hundreds of thousands or several million) therefore, the number
of false-positive can be high even for a small value of α. For example,
with α=1%, 500000 markers, and if none of them is associated with the
phenotypic trait, we can still expect 5000 “significant test”.

To avoid that, the p-values must be
[adjusted](https://en.wikipedia.org/wiki/Multiple_comparisons_problem)
according to the number of markers.

Finally, the result of a GWAS analysis is presented as a
[Manhattan\_plot](https://en.wikipedia.org/wiki/Manhattan_plot) which
shows the `-log(p-value)` on the y-axis and all the markers (order
according to their position on the chromosomes) on the x-axis. A
horizontal line represents the significance threshold and the points of
the markers above the line can be considered as being associated with
the phenotypic trait.

![screenshot of the plotly graph](README_files/manPlot.png)

GWAS-Engine provides command line tools to do GWAS analysis.

> Note: You can also check this [7min
> video](https://www.youtube.com/watch?v=sOP8WacfBM8) by Nuno Carvalho
> that explain GWAS very well.

# Build GWAS-Engine

The easiest way to use GWAS-Engine is to use the
[Docker](https://www.docker.com/) image. To build the docker image
simply do:

``` sh
git clone https://github.com/ut-biomet/GWAS-Engine
cd GWAS-Engine
docker build -t gwasengine ./
```

## Test the image

You can test that the image is build correctly by running:

``` sh
docker run --entrypoint="Rscript" gwasengine ./tests/testthat.R
```

All tests should pass.

# How to use GWAS-Engine

Basically the command to use the engine from the docker image is as
follow:

``` sh
docker run gwasengine tool --param1 valueOfParam1 --param2 valueOfParam2 ...
```

-   `tool` can be either:
    -   `gwas`: to do a gwas analysis from geno and pheno files
    -   `manplot`: to create an interactive manhathan plot from the
        results generated by `gwas`
    -   `adjresults`: to calculate the adjusted p-value from the results
        generated by `gwas`
    -   `ldplot`: to get a plot showing the Linkage disequilibrium
        between consecutive markers.

The other parameters depends of the tool and you can have the exhaustive
list of them with the command:

``` sh
docker run gwasengine
```

<details>
<summary>
Click to see outputs
</summary>

``` sh
docker run gwasengine
#> usage: gwas-engine.R [--] [--help] [--opts OPTS] [--genoFile GENOFILE]
#>        [--phenoFile PHENOFILE] [--gwasFile GWASFILE] [--outFile
#>        OUTFILE] [--trait TRAIT] [--chr CHR] [--test TEST] [--fixed
#>        FIXED] [--response RESPONSE] [--adj_method ADJ_METHOD]
#>        [--thresh_maf THRESH_MAF] [--thresh_callrate THRESH_CALLRATE]
#>        [--thresh_p THRESH_P] [--from FROM] [--to TO] fun
#> 
#> Run GWAS-Engine's tools
#> 
#> positional arguments:
#>   fun                GWAS-Engine's function you want to run: `gwas`,
#>                      `manplot`, `ldplot` or `adjresults`
#> 
#> flags:
#>   -h, --help         show this help message and exit
#> 
#> optional arguments:
#>   -x, --opts         RDS file containing argument values
#>   -g, --genoFile     [`gwas`, `ldplot`] path of the geno data file
#>                      (`.vcf` or `.vcf.gz` file)
#>   -p, --phenoFile    [`gwas`] path of the phenotypic data file (`csv`
#>                      file)
#>   --gwasFile         [`manplot`, `adjresults`] path of the gwas result
#>                      data file (json file)
#>   -o, --outFile      [`gwas`, `manplot`, `ldplot`, `adjresults`] file
#>                      where to save the results,
#>   -t, --trait        [`gwas`] name of the trait to analyze. Must be a
#>                      column name of the phenotypic file.
#>   -c, --chr          [`manplot`] name of the chromosome to show (show
#>                      all by default)
#>   --test             [`gwas`] Which test to use. Either `"score"`,
#>                      `"wald"` or `"lrt"`. For binary phenotypes, test =
#>                      `"score"` is mandatory. For more information about
#>                      this parameters see:
#>                      https://www.rdocumentation.org/packages/gaston/versions/1.4.9/topics/association.test
#>   -f, --fixed        [`gwas`] Number of Principal Components to include
#>                      in the model with fixed effect (for test =
#>                      `"wald"` or `"lrt"`). Default value is 0. For more
#>                      information about this parameters see:
#>                      https://www.rdocumentation.org/packages/gaston/versions/1.4.9/topics/association.test
#>                      [default: 0]
#>   -r, --response     [`gwas`] Either "quantitative" or "binary". Is the
#>                      trait a quantitative or a binary phenotype?
#>                      Default value is "quantitative" [default:
#>                      quantitative]
#>   -a, --adj_method   [`manplot`, `adjresults`] correction method:
#>                      "holm", "hochberg", "bonferroni", "BH", "BY",
#>                      "fdr", "none". Default: "bonferroni". (see
#>                      https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/p.adjust
#>                      for more details) [default: bonferroni]
#>   --thresh_maf       [`gwas`] Threshold for filtering markers. Only
#>                      markers with minor allele frequency > `thresh_maf`
#>                      will be kept.
#>   --thresh_callrate  [`gwas`] Threshold for filtering markers. Only
#>                      markers with a callrate > `thresh_callrate` will
#>                      be kept.
#>   --thresh_p         [`manplot`, `adjresults`] p value significant
#>                      threshold (default 0.05) [default: 0.05]
#>   --from             [`ldplot`] lower bound of the range of SNPs for
#>                      which the LD is computed
#>   --to               [`ldplot`] upper bound of the range of SNPs for
#>                      which the LD is computed
```

</details>

However, in order to let the docker container read/write files from the
host machine, you need to mount the the folders containing the files in
the container using the option `-v --volume`. Then the file path must be
specify relative to the place in the container. (See examples below).

## Examples

``` sh
docker run -v "$PWD"/data/geno/:/geno \
    -v "$PWD"/data/pheno/:/pheno \
    -v "$PWD"/readmeTemp:/out gwasengine \
    gwas \
    --genoFile "/geno/testMarkerData01.vcf.gz" \
    --phenoFile "/pheno/testPhenoData01.csv" \
    --trait "Flowering.time.at.Arkansas" \
    --test "score" \
    --fixed 0 \
    --response "quantitative" \
    --thresh_maf 0.05 \
    --thresh_callrate 0.95 \
    --outFile "/out/gwasRes.json"
```

<details>
<summary>
Click to see outputs
</summary>

``` sh
docker run -v "$PWD"/data/geno/:/geno \
    -v "$PWD"/data/pheno/:/pheno \
    -v "$PWD"/readmeTemp:/out gwasengine \
    gwas \
    --genoFile "/geno/testMarkerData01.vcf.gz" \
    --phenoFile "/pheno/testPhenoData01.csv" \
    --trait "Flowering.time.at.Arkansas" \
    --test "score" \
    --fixed 0 \
    --response "quantitative" \
    --thresh_maf 0.05 \
    --thresh_callrate 0.95 \
    --outFile "/out/gwasRes.json"
#> 2021-09-08 10:35:10 call to GWAS-Engine.R:
#>      help = FALSE
#>      opts = NA
#>      genoFile = /geno/testMarkerData01.vcf.gz
#>      phenoFile = /pheno/testPhenoData01.csv
#>      gwasFile = NA
#>      outFile = /out/gwasRes.json
#>      trait = Flowering.time.at.Arkansas
#>      chr = NA
#>      test = score
#>      fixed = 0
#>      response = quantitative
#>      adj_method = bonferroni
#>      thresh_maf = 0.05
#>      thresh_callrate = 0.95
#>      thresh_p = 0.05
#>      from = NA
#>      to = NA
#>      fun = gwas
#> 2021-09-08 10:35:11 - r-run_gwas(): Get data ...
#> 2021-09-08 10:35:11 - r-readData(): get geno data ...
#> 2021-09-08 10:35:11 - r-readGenoData(): Check file extention ... 
#> 2021-09-08 10:35:11 - r-readGenoData(): Read geno file ... 
#> ped stats and snps stats have been set. 
#> 'p' has been set. 
#> 'mu' and 'sigma' have been set.
#> 2021-09-08 10:35:11 - r-readGenoData(): Read geno file DONE 
#> 2021-09-08 10:35:11 - r-readGenoData(): DONE, return output.
#> 2021-09-08 10:35:11 - r-readData(): get geno data DONE
#> 2021-09-08 10:35:11 - r-readData(): get pheno data ...
#> 2021-09-08 10:35:11 - r-readPhenoData(): Read phenotypic file ... 
#> 2021-09-08 10:35:11 - r-readPhenoData(): Read phenotypic file DONE 
#> 2021-09-08 10:35:11 - r-readPhenoData(): DONE, return output.
#> 2021-09-08 10:35:11 - r-readData(): get pheno data DONE
#> 2021-09-08 10:35:11 - r-readData(): prepare data ...
#> 2021-09-08 10:35:11 - r-prepareData(): Remove from geno data individuals that are not in phenotypic data-set ...
#> 2021-09-08 10:35:12 - r-prepareData(): Remove from geno data individuals that are not in phenotypic data-set DONE
#> 2021-09-08 10:35:12 - r-prepareData(): reorder matrix ...
#> 2021-09-08 10:35:12 - r-prepareData(): reorder matrix DONE
#> 2021-09-08 10:35:12 - r-prepareData(): remove monomorphic geno ...
#> 2021-09-08 10:35:12 - r-prepareData(): remove monomorphic geno DONE
#> 2021-09-08 10:35:12 - r-prepareData(): calculate genetic relatinoal matrix ...
#> 2021-09-08 10:35:12 - r-prepareData(): calculate genetic relatinoal matrix DONE
#> 2021-09-08 10:35:12 - r-prepareData(): DONE, return output.
#> 2021-09-08 10:35:12 - r-readData(): prepare data DONE
#> 2021-09-08 10:35:12 - r-readData(): DONE, return output.
#> 2021-09-08 10:35:12 - r-run_gwas(): Get data DONE
#> 2021-09-08 10:35:12 - r-run_gwas(): GWAS analysis ...
#> 2021-09-08 10:35:13 - r-gwas(): Check inputs ...
#> 2021-09-08 10:35:13 - r-gwas(): Check inputs DONE
#> 2021-09-08 10:35:13 - r-gwas(): aggregate data in bed matrix ...
#> 2021-09-08 10:35:13 - r-gwas(): aggregate data in bed matrix DONE
#> 2021-09-08 10:35:13 - r-gwas(): remove individuals with missing phenotypic values ...
#> 2021-09-08 10:35:13 - r-gwas(): remove samples with missing phenotypic values DONE
#> 2021-09-08 10:35:13 - r-gwas(): filter SNPs ...
#> 2021-09-08 10:35:13 - r-gwas(): filter SNPs DONE
#> 2021-09-08 10:35:13 - r-gwas(): fit model ...
#> [Iteration 1] theta = 78.0648 39.8993
#> [Iteration 1] log L = -1000.94
#> [Iteration 1] AI-REML update
#> [Iteration 1] ||gradient|| = 0.403712
#> [Iteration 2] theta = 19.3247 82.1855
#> [Iteration 2] log L = -991.683
#> [Iteration 2] AI-REML update
#> [Iteration 2] ||gradient|| = 1.19645
#> [Iteration 3] theta = 25.9637 94.4355
#> [Iteration 3] log L = -985.452
#> [Iteration 3] AI-REML update
#> [Iteration 3] ||gradient|| = 0.200586
#> [Iteration 4] theta = 29.0151 95.1047
#> [Iteration 4] log L = -985.125
#> [Iteration 4] AI-REML update
#> [Iteration 4] ||gradient|| = 0.0176329
#> [Iteration 5] theta = 29.5483  94.409
#> [Iteration 5] log L = -985.121
#> [Iteration 5] AI-REML update
#> [Iteration 5] ||gradient|| = 0.0013669
#> [Iteration 6] theta = 29.6082 94.2813
#> [Iteration 6] log L = -985.121
#> [Iteration 6] AI-REML update
#> [Iteration 6] ||gradient|| = 0.000156204
#> [Iteration 7] theta = 29.6154 94.2651
#> [Iteration 7] log L = -985.121
#> [Iteration 7] AI-REML update
#> [Iteration 7] ||gradient|| = 1.8962e-05
#> [Iteration 8] theta = 29.6163 94.2631
#> [Iteration 8] log L = -985.121
#> [Iteration 8] AI-REML update
#> [Iteration 8] ||gradient|| = 2.31737e-06
#> 2021-09-08 10:35:13 - r-gwas(): fit model DONE
#> 2021-09-08 10:35:13 - r-gwas(): DONE, return output.
#> 2021-09-08 10:35:13 - r-run_gwas(): GWAS analysis DONE
#> 2021-09-08 10:35:13 - r-run_gwas(): Save metadata ...
#> 2021-09-08 10:35:13 - r-run_gwas(): Save metadata DONE
#> 2021-09-08 10:35:13 - r-run_gwas(): Save results ...
#> 2021-09-08 10:35:13 - r-saveGWAS(): Check file ...
#> 2021-09-08 10:35:13 - r-saveGWAS(): Warning: "file" directory already exists. This file will be overwritten.
#> 2021-09-08 10:35:13 - r-saveGWAS(): Check file DONE
#> 2021-09-08 10:35:13 - r-run_gwas(): Save results DONE
```

</details>

### Draw Manhattan Plot

``` sh
docker run -v "$PWD"/readmeTemp:/files gwasengine \
    manplot \
    --gwasFile "/files/gwasRes.json" \
    --adj_method "bonferroni" \
    --thresh_p 0.05 \
    --outFile "/files/manPlot.html"
```

<details>
<summary>
Click to see outputs
</summary>

``` sh
docker run -v "$PWD"/readmeTemp:/files gwasengine \
    manplot \
    --gwasFile "/files/gwasRes.json" \
    --adj_method "bonferroni" \
    --thresh_p 0.05 \
    --outFile "/files/manPlot.html"
#> 2021-09-08 10:35:17 call to GWAS-Engine.R:
#>      help = FALSE
#>      opts = NA
#>      genoFile = NA
#>      phenoFile = NA
#>      gwasFile = /files/gwasRes.json
#>      outFile = /files/manPlot.html
#>      trait = NA
#>      chr = NA
#>      test = NA
#>      fixed = 0
#>      response = quantitative
#>      adj_method = bonferroni
#>      thresh_maf = NA
#>      thresh_callrate = NA
#>      thresh_p = 0.05
#>      from = NA
#>      to = NA
#>      fun = manplot
#> 2021-09-08 10:35:17 - r-draw_manhattanPlot(): Get data ...
#> 2021-09-08 10:35:17 - r-readGWAS(): Read result file ... 
#> 2021-09-08 10:35:17 - r-readGWAS(): Read result file DONE 
#> 2021-09-08 10:35:17 - r-readGWAS(): Convert Json to data.frame ... 
#> 2021-09-08 10:35:18 - r-readGWAS(): Convert Json to data.frame DONE 
#> 2021-09-08 10:35:18 - r-readGWAS(): DONE, return output.
#> 2021-09-08 10:35:18 - r-draw_manhattanPlot(): Get data DONE
#> 2021-09-08 10:35:18 - r-draw_manhattanPlot(): Draw Manhattan Plot ...
#> 2021-09-08 10:35:18 - r-manPlot(): Check chromosome name ...
#> 2021-09-08 10:35:18 - r-manPlot(): Check chromosome name DONE
#> 2021-09-08 10:35:18 - r-manPlot(): Adjust p-values ...
#> 2021-09-08 10:35:18 - r-adjustPval(): Check adj_method ...
#> 2021-09-08 10:35:18 - r-adjustPval(): Check adj_method DONE
#> 2021-09-08 10:35:18 - r-adjustPval(): Check p values ...
#> 2021-09-08 10:35:18 - r-adjustPval(): Check p values DONE
#> 2021-09-08 10:35:18 - r-adjustPval(): Adjust p-values ...
#> 2021-09-08 10:35:18 - r-adjustPval(): Adjust p-values DONE
#> 2021-09-08 10:35:18 - r-adjustPval(): Adjust threshold ...
#> 2021-09-08 10:35:18 - r-adjustPval(): Adjust threshold DONE
#> 2021-09-08 10:35:18 - r-adjustPval(): DONE, return output
#> 2021-09-08 10:35:18 - r-manPlot(): Adjust p-values DONE
#> 2021-09-08 10:35:19 - r-manPlot(): DONE, return output
#> 2021-09-08 10:35:19 - r-draw_manhattanPlot(): Draw Manhattan Plot DONE
#> 2021-09-08 10:35:19 - r-draw_manhattanPlot(): Save results ...
#> 2021-09-08 10:35:29 - r-draw_manhattanPlot(): Save results DONE
```

![screenshot of the plotly graph](README_files/manPlot.png)

</details>

### Adjust p-values

``` sh
docker run -v "$PWD"/readmeTemp:/files gwasengine \
    adjresults \
    --gwasFile "/files/gwasRes.json" \
    --adj_method "bonferroni" \
    --outFile "/files/adjRes.json"
```

<details>
<summary>
Click to see outputs
</summary>

``` sh
docker run -v "$PWD"/readmeTemp:/files gwasengine \
    adjresults \
    --gwasFile "/files/gwasRes.json" \
    --adj_method "bonferroni" \
    --outFile "/files/adjRes.json"
#> 2021-09-08 10:35:32 call to GWAS-Engine.R:
#>      help = FALSE
#>      opts = NA
#>      genoFile = NA
#>      phenoFile = NA
#>      gwasFile = /files/gwasRes.json
#>      outFile = /files/adjRes.json
#>      trait = NA
#>      chr = NA
#>      test = NA
#>      fixed = 0
#>      response = quantitative
#>      adj_method = bonferroni
#>      thresh_maf = NA
#>      thresh_callrate = NA
#>      thresh_p = 0.05
#>      from = NA
#>      to = NA
#>      fun = adjresults
#> 2021-09-08 10:35:32 - r-run_resAdjustment(): Get data ...
#> 2021-09-08 10:35:32 - r-readGWAS(): Read result file ... 
#> 2021-09-08 10:35:32 - r-readGWAS(): Read result file DONE 
#> 2021-09-08 10:35:32 - r-readGWAS(): Convert Json to data.frame ... 
#> 2021-09-08 10:35:33 - r-readGWAS(): Convert Json to data.frame DONE 
#> 2021-09-08 10:35:33 - r-readGWAS(): DONE, return output.
#> 2021-09-08 10:35:33 - r-run_resAdjustment(): Get data DONE
#> 2021-09-08 10:35:33 - r-run_resAdjustment(): Adjust p-values ...
#> 2021-09-08 10:35:33 - r-adjustPval(): Check adj_method ...
#> 2021-09-08 10:35:33 - r-adjustPval(): Check adj_method DONE
#> 2021-09-08 10:35:33 - r-adjustPval(): Check p values ...
#> 2021-09-08 10:35:33 - r-adjustPval(): Check p values DONE
#> 2021-09-08 10:35:33 - r-adjustPval(): Adjust p-values ...
#> 2021-09-08 10:35:33 - r-adjustPval(): Adjust p-values DONE
#> 2021-09-08 10:35:33 - r-adjustPval(): DONE, return output
#> 2021-09-08 10:35:33 - r-run_resAdjustment(): Adjust p-values DONE
#> 2021-09-08 10:35:33 - r-run_resAdjustment(): Save results ...
#> 2021-09-08 10:35:33 - r-saveGWAS(): Check file ...
#> 2021-09-08 10:35:33 - r-saveGWAS(): Warning: "file" directory already exists. This file will be overwritten.
#> 2021-09-08 10:35:33 - r-saveGWAS(): Check file DONE
#> 2021-09-08 10:35:33 - r-run_resAdjustment(): Save results DONE
```

</details>

### Draw LD plot

``` sh
docker run -v "$PWD"/data/geno/:/geno \
    -v "$PWD"/readmeTemp:/out gwasengine \
    ldplot \
    --genoFile "/geno/testMarkerData01.vcf.gz" \
    --from 42 \
    --to 62 \
    --outFile "/out/ldplot.png"
```

<details>
<summary>
Click to see outputs
</summary>

``` sh
docker run -v "$PWD"/data/geno/:/geno \
    -v "$PWD"/readmeTemp:/out gwasengine \
    ldplot \
    --genoFile "/geno/testMarkerData01.vcf.gz" \
    --from 42 \
    --to 62 \
    --outFile "/out/ldplot.png"
#> 2021-09-08 10:35:37 call to GWAS-Engine.R:
#>      help = FALSE
#>      opts = NA
#>      genoFile = /geno/testMarkerData01.vcf.gz
#>      phenoFile = NA
#>      gwasFile = NA
#>      outFile = /out/ldplot.png
#>      trait = NA
#>      chr = NA
#>      test = NA
#>      fixed = 0
#>      response = quantitative
#>      adj_method = bonferroni
#>      thresh_maf = NA
#>      thresh_callrate = NA
#>      thresh_p = 0.05
#>      from = 42
#>      to = 62
#>      fun = ldplot
#> 2021-09-08 10:35:37 - r-draw_ldPlot(): Get data ...
#> 2021-09-08 10:35:37 - r-readGenoData(): Check file extention ... 
#> 2021-09-08 10:35:37 - r-readGenoData(): Read geno file ... 
#> ped stats and snps stats have been set. 
#> 'p' has been set. 
#> 'mu' and 'sigma' have been set.
#> 2021-09-08 10:35:38 - r-readGenoData(): Read geno file DONE 
#> 2021-09-08 10:35:38 - r-readGenoData(): DONE, return output.
#> 2021-09-08 10:35:38 - r-draw_ldPlot(): Get data DONE
#> 2021-09-08 10:35:38 - r-draw_ldPlot(): Draw LD Plot ...
#> 2021-09-08 10:35:38 - r-LDplot(): Check "from" and "to" format ...
#> 2021-09-08 10:35:38 - r-LDplot(): Check "from" and "to" format DONE
#> 2021-09-08 10:35:38 - r-LDplot(): Check "from" < "to"...
#> 2021-09-08 10:35:38 - r-LDplot(): Check "from" < "to" DONE
#> 2021-09-08 10:35:38 - r-LDplot(): Check number of SNP < 50...
#> 2021-09-08 10:35:38 - r-LDplot(): Check number of SNP < 50 DONE
#> 2021-09-08 10:35:38 - r-LDplot(): Check file ...
#> 2021-09-08 10:35:38 - r-LDplot(): Warning: "file" directory already exists. This file will be overwritten.
#> 2021-09-08 10:35:38 - r-LDplot(): Check file DONE
#> 2021-09-08 10:35:38 - r-LDplot(): Compute LD ...
#> 2021-09-08 10:35:38 - r-LDplot(): Compute LD DONE
#> 2021-09-08 10:35:38 - r-LDplot(): Create LD plot ...
#> 2021-09-08 10:35:38 - r-LDplot(): Create create file: /out/ldplot.png
#> 2021-09-08 10:35:39 - r-LDplot(): Create LD plot DONE
#> 2021-09-08 10:35:39 - r-LDplot(): DONE, return output
#> 2021-09-08 10:35:39 - r-draw_ldPlot(): Draw LD Plot DONE
```

``` r
library(png)
img <- readPNG("readmeTemp/ldplot.png")
grid::grid.raster(img)
```

![](README_files/figure-gfm/unnamed-chunk-7-1.png)<!-- -->

</details>

# For `R` developpers:

This GWAS-Engine is written in `R`, therefore it can also be run
directly from your computer (without docker) if you have all the
dependencies installed. To do so you can execute the `gwas-engine.R`
file with the `Rscript` command.

For example from a bash terminal:

``` sh
Rscript ./gwas-engine.R \
    gwas \
    --genoFile "data/geno/testMarkerData01.vcf.gz" \
    --phenoFile "data/pheno/testPhenoData01.csv" \
    --trait "Flowering.time.at.Arkansas" \
    --test "score" \
    --fixed 0 \
    --response "quantitative" \
    --thresh_maf 0.05 \
    --thresh_callrate 0.95 \
    --outFile "readmeTemp/gwasRes.json"
```

## Dependencies

Package dependencies are managed using
[Renv](https://rstudio.github.io/renv/articles/renv.html). To install
dependencies do in a R console: `renv::restore()`. (you need
`/path/to/GWAS-Engine` as working directory).

``` r
# required packages for using GWAS-Engine:
library(argparser) # Command-Line Argument Parser
library(gaston) # for many functions
#> Loading required package: Rcpp
#> Loading required package: RcppParallel
#> 
#> Attaching package: 'RcppParallel'
#> The following object is masked from 'package:Rcpp':
#> 
#>     LdFlags
#> Gaston set number of threads to 4. Use setThreadOptions() to modify this.
#> 
#> Attaching package: 'gaston'
#> The following object is masked from 'package:stats':
#> 
#>     sigma
#> The following objects are masked from 'package:base':
#> 
#>     cbind, rbind
library(jsonlite) # manage json format
library(manhattanly) # manhattan plot using plotly
#> See example usage at http://sahirbhatnagar.com/manhattanly/
library(R6) # R object oriented
```

## GWAS-Engine as a R library

It is possible to use GWAS-Engine as a library to be used in other R
projects.

To do so, you can load all the code of this engine using

``` r
sapply(FUN = source,
       X = list.files("/path/to/GWAS-Engine/src", pattern = ".R$", full.names = T))
```

### Main functions

GWAS-Engine provide 4 main R functions similar to what can be done with
command line arguments.

They take as inputs file paths and other specific argument and write the
results in output files:

``` r
gwas_results <- run_gwas(genoFile = "data/geno/testMarkerData01.vcf.gz",
                         phenoFile = "data/pheno/testPhenoData01.csv",
                         genoUrl = NULL,
                         phenoUrl = NULL,
                         trait = "Flowering.time.at.Arkansas",
                         test = "score",
                         fixed = 0,
                         response = "quantitative",
                         thresh_maf = 0.05,
                         thresh_callrate = 0.95,
                         outFile = tempfile(fileext = ".json"))
#> 2021-09-08 19:35:41 - r-run_gwas(): Get data ...
#> 2021-09-08 19:35:41 - r-readData(): get geno data ...
#> 2021-09-08 19:35:41 - r-readGenoData(): Check file extention ... 
#> 2021-09-08 19:35:41 - r-readGenoData(): Read geno file ... 
#> ped stats and snps stats have been set. 
#> 'p' has been set. 
#> 'mu' and 'sigma' have been set.
#> 2021-09-08 19:35:42 - r-readGenoData(): Read geno file DONE 
#> 2021-09-08 19:35:42 - r-readGenoData(): DONE, return output.
#> 2021-09-08 19:35:42 - r-readData(): get geno data DONE
#> 2021-09-08 19:35:42 - r-readData(): get pheno data ...
#> 2021-09-08 19:35:42 - r-readPhenoData(): Read phenotypic file ... 
#> 2021-09-08 19:35:42 - r-readPhenoData(): Read phenotypic file DONE 
#> 2021-09-08 19:35:42 - r-readPhenoData(): DONE, return output.
#> 2021-09-08 19:35:42 - r-readData(): get pheno data DONE
#> 2021-09-08 19:35:42 - r-readData(): prepare data ...
#> 2021-09-08 19:35:42 - r-prepareData(): Remove from geno data individuals that are not in phenotypic data-set ...
#> 2021-09-08 19:35:42 - r-prepareData(): Remove from geno data individuals that are not in phenotypic data-set DONE
#> 2021-09-08 19:35:42 - r-prepareData(): reorder matrix ...
#> 2021-09-08 19:35:42 - r-prepareData(): reorder matrix DONE
#> 2021-09-08 19:35:42 - r-prepareData(): remove monomorphic geno ...
#> 2021-09-08 19:35:42 - r-prepareData(): remove monomorphic geno DONE
#> 2021-09-08 19:35:42 - r-prepareData(): calculate genetic relatinoal matrix ...
#> 2021-09-08 19:35:43 - r-prepareData(): calculate genetic relatinoal matrix DONE
#> 2021-09-08 19:35:43 - r-prepareData(): DONE, return output.
#> 2021-09-08 19:35:43 - r-readData(): prepare data DONE
#> 2021-09-08 19:35:43 - r-readData(): DONE, return output.
#> 2021-09-08 19:35:43 - r-run_gwas(): Get data DONE
#> 2021-09-08 19:35:43 - r-run_gwas(): GWAS analysis ...
#> 2021-09-08 19:35:43 - r-gwas(): Check inputs ...
#> 2021-09-08 19:35:43 - r-gwas(): Check inputs DONE
#> 2021-09-08 19:35:43 - r-gwas(): aggregate data in bed matrix ...
#> 2021-09-08 19:35:43 - r-gwas(): aggregate data in bed matrix DONE
#> 2021-09-08 19:35:43 - r-gwas(): remove individuals with missing phenotypic values ...
#> 2021-09-08 19:35:43 - r-gwas(): remove samples with missing phenotypic values DONE
#> 2021-09-08 19:35:43 - r-gwas(): filter SNPs ...
#> 2021-09-08 19:35:43 - r-gwas(): filter SNPs DONE
#> 2021-09-08 19:35:43 - r-gwas(): fit model ...
#> [Iteration 1] theta = 78.0648 39.8993
#> [Iteration 1] log L = -1000.94
#> [Iteration 1] AI-REML update
#> [Iteration 1] ||gradient|| = 0.403711
#> [Iteration 2] theta = 19.3247 82.1855
#> [Iteration 2] log L = -991.683
#> [Iteration 2] AI-REML update
#> [Iteration 2] ||gradient|| = 1.19645
#> [Iteration 3] theta = 25.9637 94.4354
#> [Iteration 3] log L = -985.452
#> [Iteration 3] AI-REML update
#> [Iteration 3] ||gradient|| = 0.200585
#> [Iteration 4] theta = 29.0152 95.1047
#> [Iteration 4] log L = -985.126
#> [Iteration 4] AI-REML update
#> [Iteration 4] ||gradient|| = 0.0176329
#> [Iteration 5] theta = 29.5483  94.409
#> [Iteration 5] log L = -985.121
#> [Iteration 5] AI-REML update
#> [Iteration 5] ||gradient|| = 0.0013669
#> [Iteration 6] theta = 29.6082 94.2812
#> [Iteration 6] log L = -985.121
#> [Iteration 6] AI-REML update
#> [Iteration 6] ||gradient|| = 0.000156203
#> [Iteration 7] theta = 29.6154  94.265
#> [Iteration 7] log L = -985.121
#> [Iteration 7] AI-REML update
#> [Iteration 7] ||gradient|| = 1.89619e-05
#> [Iteration 8] theta = 29.6163 94.2631
#> [Iteration 8] log L = -985.121
#> [Iteration 8] AI-REML update
#> [Iteration 8] ||gradient|| = 2.31736e-06
#> 2021-09-08 19:35:43 - r-gwas(): fit model DONE
#> 2021-09-08 19:35:43 - r-gwas(): DONE, return output.
#> 2021-09-08 19:35:43 - r-run_gwas(): GWAS analysis DONE
#> 2021-09-08 19:35:43 - r-run_gwas(): Save metadata ...
#> 2021-09-08 19:35:43 - r-run_gwas(): Save metadata DONE
#> 2021-09-08 19:35:43 - r-run_gwas(): Save results ...
#> 2021-09-08 19:35:43 - r-saveGWAS(): Check file ...
#> 2021-09-08 19:35:43 - r-saveGWAS(): Check file DONE
#> 2021-09-08 19:35:44 - r-run_gwas(): Save results DONE
gwas_results$file
#> [1] "/tmp/RtmpqCRIwW/file6d0113d803f2.json"
substr(gwas_results$gwasRes, start=1, stop=500)
#> [
#>   {
#>     "chr": "1",
#>     "pos": 9563,
#>     "id": "SNP-1.8562.",
#>     "A1": "A",
#>     "A2": "T",
#>     "freqA2": 0.1359,
#>     "score": 3.6664,
#>     "p": 0.0555
#>   },
#>   {
#>     "chr": "1",
#>     "pos": 25922,
#>     "id": "SNP-1.24921.",
#>     "A1": "C",
#>     "A2": "T",
#>     "freqA2": 0.1254,
#>     "score": 1.3876,
#>     "p": 0.2388
#>   },
#>   {
#>     "chr": "1",
#>     "pos": 26254,
#>     "id": "SNP-1.25253.",
#>     "A1": "A",
#>     "A2": "T",
#>     "freqA2": 0.2935,
#>     "score": 2.9466,
#>     "p": 0.0861
#>   },
#>   {
#>     "chr": "1",
#>     "p
```

``` r
p <- draw_manhattanPlot(gwasFile = gwas_results$file,
                        gwasUrl = NULL,
                        adj_method = "bonferroni",
                        thresh_p = 0.05,
                        chr = NA,
                        outFile = tempfile(fileext = ".html"))
#> 2021-09-08 19:35:44 - r-draw_manhattanPlot(): Get data ...
#> 2021-09-08 19:35:44 - r-readGWAS(): Read result file ... 
#> 2021-09-08 19:35:44 - r-readGWAS(): Read result file DONE 
#> 2021-09-08 19:35:44 - r-readGWAS(): Convert Json to data.frame ... 
#> 2021-09-08 19:35:44 - r-readGWAS(): Convert Json to data.frame DONE 
#> 2021-09-08 19:35:44 - r-readGWAS(): DONE, return output.
#> 2021-09-08 19:35:44 - r-draw_manhattanPlot(): Get data DONE
#> 2021-09-08 19:35:44 - r-draw_manhattanPlot(): Draw Manhattan Plot ...
#> 2021-09-08 19:35:44 - r-manPlot(): Check chromosome name ...
#> 2021-09-08 19:35:44 - r-manPlot(): Check chromosome name DONE
#> 2021-09-08 19:35:44 - r-manPlot(): Adjust p-values ...
#> 2021-09-08 19:35:44 - r-adjustPval(): Check adj_method ...
#> 2021-09-08 19:35:44 - r-adjustPval(): Check adj_method DONE
#> 2021-09-08 19:35:44 - r-adjustPval(): Check p values ...
#> 2021-09-08 19:35:44 - r-adjustPval(): Check p values DONE
#> 2021-09-08 19:35:44 - r-adjustPval(): Adjust p-values ...
#> 2021-09-08 19:35:44 - r-adjustPval(): Adjust p-values DONE
#> 2021-09-08 19:35:44 - r-adjustPval(): Adjust threshold ...
#> 2021-09-08 19:35:44 - r-adjustPval(): Adjust threshold DONE
#> 2021-09-08 19:35:44 - r-adjustPval(): DONE, return output
#> 2021-09-08 19:35:44 - r-manPlot(): Adjust p-values DONE
#> 2021-09-08 19:35:45 - r-manPlot(): DONE, return output
#> 2021-09-08 19:35:45 - r-draw_manhattanPlot(): Draw Manhattan Plot DONE
#> 2021-09-08 19:35:45 - r-draw_manhattanPlot(): Save results ...
#> 2021-09-08 19:35:49 - r-draw_manhattanPlot(): Save results DONE
```

``` r
gwas_adj <- run_resAdjustment(gwasFile = gwas_results$file,
                              gwasUrl = NULL,
                              adj_method = "bonferroni",
                              outFile = tempfile(fileext = ".json"))
#> 2021-09-08 19:35:49 - r-run_resAdjustment(): Get data ...
#> 2021-09-08 19:35:49 - r-readGWAS(): Read result file ... 
#> 2021-09-08 19:35:49 - r-readGWAS(): Read result file DONE 
#> 2021-09-08 19:35:49 - r-readGWAS(): Convert Json to data.frame ... 
#> 2021-09-08 19:35:49 - r-readGWAS(): Convert Json to data.frame DONE 
#> 2021-09-08 19:35:49 - r-readGWAS(): DONE, return output.
#> 2021-09-08 19:35:49 - r-run_resAdjustment(): Get data DONE
#> 2021-09-08 19:35:49 - r-run_resAdjustment(): Adjust p-values ...
#> 2021-09-08 19:35:49 - r-adjustPval(): Check adj_method ...
#> 2021-09-08 19:35:49 - r-adjustPval(): Check adj_method DONE
#> 2021-09-08 19:35:49 - r-adjustPval(): Check p values ...
#> 2021-09-08 19:35:49 - r-adjustPval(): Check p values DONE
#> 2021-09-08 19:35:49 - r-adjustPval(): Adjust p-values ...
#> 2021-09-08 19:35:49 - r-adjustPval(): Adjust p-values DONE
#> 2021-09-08 19:35:49 - r-adjustPval(): DONE, return output
#> 2021-09-08 19:35:49 - r-run_resAdjustment(): Adjust p-values DONE
#> 2021-09-08 19:35:49 - r-run_resAdjustment(): Save results ...
#> 2021-09-08 19:35:49 - r-saveGWAS(): Check file ...
#> 2021-09-08 19:35:49 - r-saveGWAS(): Check file DONE
#> 2021-09-08 19:35:50 - r-run_resAdjustment(): Save results DONE
substr(gwas_adj$gwasAdjusted, start=1, stop=500)
#> [
#>   {
#>     "chr": "1",
#>     "pos": 9563,
#>     "id": "SNP-1.8562.",
#>     "A1": "A",
#>     "A2": "T",
#>     "freqA2": 0.1359,
#>     "score": 3.6664,
#>     "p": 0.0555,
#>     "p_adj": 1
#>   },
#>   {
#>     "chr": "1",
#>     "pos": 25922,
#>     "id": "SNP-1.24921.",
#>     "A1": "C",
#>     "A2": "T",
#>     "freqA2": 0.1254,
#>     "score": 1.3876,
#>     "p": 0.2388,
#>     "p_adj": 1
#>   },
#>   {
#>     "chr": "1",
#>     "pos": 26254,
#>     "id": "SNP-1.25253.",
#>     "A1": "A",
#>     "A2": "T",
#>     "freqA2": 0.2935,
#>     "score": 2.9466,
#>     "p": 0.0861
```

``` r
imgFile <- draw_ldPlot(genoFile = "data/geno/testMarkerData01.vcf.gz",
                       genoUrl = NULL,
                       from = 42,
                       to = 62,
                       outFile = tempfile(fileext = ".png")) 
#> 2021-09-08 19:35:50 - r-draw_ldPlot(): Get data ...
#> 2021-09-08 19:35:50 - r-readGenoData(): Check file extention ... 
#> 2021-09-08 19:35:50 - r-readGenoData(): Read geno file ... 
#> ped stats and snps stats have been set. 
#> 'p' has been set. 
#> 'mu' and 'sigma' have been set.
#> 2021-09-08 19:35:50 - r-readGenoData(): Read geno file DONE 
#> 2021-09-08 19:35:50 - r-readGenoData(): DONE, return output.
#> 2021-09-08 19:35:50 - r-draw_ldPlot(): Get data DONE
#> 2021-09-08 19:35:50 - r-draw_ldPlot(): Draw LD Plot ...
#> 2021-09-08 19:35:50 - r-LDplot(): Check "from" and "to" format ...
#> 2021-09-08 19:35:50 - r-LDplot(): Check "from" and "to" format DONE
#> 2021-09-08 19:35:50 - r-LDplot(): Check "from" < "to"...
#> 2021-09-08 19:35:50 - r-LDplot(): Check "from" < "to" DONE
#> 2021-09-08 19:35:50 - r-LDplot(): Check number of SNP < 50...
#> 2021-09-08 19:35:50 - r-LDplot(): Check number of SNP < 50 DONE
#> 2021-09-08 19:35:50 - r-LDplot(): Check file ...
#> 2021-09-08 19:35:50 - r-LDplot(): Check file DONE
#> 2021-09-08 19:35:50 - r-LDplot(): Compute LD ...
#> 2021-09-08 19:35:50 - r-LDplot(): Compute LD DONE
#> 2021-09-08 19:35:50 - r-LDplot(): Create LD plot ...
#> 2021-09-08 19:35:50 - r-LDplot(): Create create file: /tmp/RtmpqCRIwW/file6d0115093d775.png
#> 2021-09-08 19:35:51 - r-LDplot(): Create LD plot DONE
#> 2021-09-08 19:35:51 - r-LDplot(): DONE, return output
#> 2021-09-08 19:35:51 - r-draw_ldPlot(): Draw LD Plot DONE
```

> See: [src/mainFunctions.R](src/mainFunctions.R)

### Other functions

Other R functions used by the engine are defined in files:

-   [src/gwas.R](src/gwas.R)
-   [src/readWriteData.R](src/readWriteData.R)
-   [src/plot.R](src/plot.R)

For example:

#### Load data

Example data are stored in the `data` folder.

`readData` read and prepare the data for GWAS analysis

``` r
data <- readData(genoFile = "data/geno/testMarkerData01.vcf",
                 phenoFile = "data/pheno/testPhenoData01.csv")
```

For online data, they can be loaded using:

``` r
data <- downloadData(genoUrl = "url/to/geno_data.vcf",
                     phenoUrl = "url/to/pheno_data.csv")
```

> See: [src/readWriteData.R](src/readWriteData.R)

#### GWAS

The `gwas` function return a `data.frame`.

``` r
gwasRes <- gwas(data = data,
                trait = "Flowering.time.at.Arkansas",
                test = "score",
                fixed = 0,
                response = "quantitative",
                thresh_maf = 0.05,
                thresh_callrate = 0.95)
head(gwasRes)
```

Save GWAS result in a `.json` file.

``` r
(file <- saveGWAS(gwasRes, metadata = "README"))
cat(paste(readLines(file, n=20), collapse = "\n"))
```

#### Manhattan plot

This function generates a “plotly” graph.

``` r
p <- manPlot(gwas = gwasRes,
             adj_method = "bonferroni",
             thresh_p = 0.05,
             chr = NA,
             title = "Readme Example")
```

<!-- ![screenshot of the plotly graph](README_files/manPlot2.png) -->

#### LD PLot

Compute *r*<sup>2</sup> Linkage Disequilibrium (LD) between given SNPs
and save a plot in a temporary PNG file.

``` r
gDta <- readGenoData("data/geno/testMarkerData01.vcf.gz")
(imgFile <- LDplot(geno = gDta,
                   from = 1,
                   to = 20,
                   file = tempfile(fileext = ".png")))
```

``` r
img <- readPNG(imgFile)
grid::grid.raster(img)
```

### Functions documentations

All the functions of this engine use `roxygen2` for documentation. You
can generate a markdown documentation located in
[`./doc/README.md`](doc/README.md) using the function `writeDoc`
(defined in [`./src/utils.R`](src/utils.R)):

``` r
writeDoc(srcDir = "src",
         docDir = "doc")
```

# Tests

The R package `testthat` is needed to run the unit tests. To run the
unit tests of this engine you can use the command:

``` sh
Rscript ./tests/testthat.R
```

# Data references

Some example data are available in the folder [data](data/). This folder
also includes examples of output files.

The genotypic and phenotypic data used as example come from:

> Keyan Zhao, Chih-Wei Tung, Georgia C. Eizenga, Mark H. Wright, M.
> Liakat Ali, Adam H. Price, Gareth J. Norton, M. Rafiqul Islam, Andy
> Reynolds, Jason Mezey, Anna M. McClung, Carlos D. Bustamante & Susan
> R. McCouch (2011). [Genome-wide association mapping reveals a rich
> genetic architecture of complex traits in *Oryza
> sativa*.](http://www.nature.com/ncomms/journal/v2/n9/full/ncomms1467.html)
> Nat Comm 2:467 \| DOI: 10.1038/ncomms1467, Published Online 13 Sep
> 2011.
